{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3ca51dec",
   "metadata": {},
   "source": [
    "# ATAC-STARR regulatory region calling script"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54dc0d5c",
   "metadata": {},
   "source": [
    "## Introduction\n",
    "\n",
    "I made a python script to call active and silent regions from ATAC-STARR data. Below I will test this script. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df8cd52a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "The following have been reloaded with a version change:\n",
      "  1) Anaconda3/5.0.1 => Anaconda3/2021.05\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#set up\n",
    "ml Anaconda3\n",
    "source activate /home/hansetj1/.conda/envs/jupyter\n",
    "cd activity_analysis/call_active_peaks"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31d668d2",
   "metadata": {},
   "source": [
    "## Test help functionality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6bd632a5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "python3: can't open file 'call_ATAC-STARR_regulatory-regions.py': [Errno 2] No such file or directory\n"
     ]
    },
    {
     "ename": "",
     "evalue": "2",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "#Test help functionality\n",
    "python3 call_ATAC-STARR_regulatory-regions.py --help"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77fef0ac",
   "metadata": {},
   "source": [
    "## Test complete run-through on a region subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ea14656e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(jupyter) (jupyter) (jupyter) (jupyter) (jupyter) PARAMETERS:\n",
      "ChrAcc Peaks File: /data/hodges_lab/ATAC-STARR_V2/data/accessibility/genrich_ATAC-like_peaks/GM12878inGM12878_DNA_genrich_Test-set.narrowPeak\n",
      "DNA Files: /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_DNA_Rep1.filtered.n-sorted.bam /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_DNA_Rep2.filtered.n-sorted.bam\n",
      "RNA Files: /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_RNA_Rep2.filtered.n-sorted.bam\n",
      "Output Directory: /data/hodges_lab/ATAC-STARR_V2/data/activity/GM12878inGM12878\n",
      "False-Discovery Rate: 0.1\n",
      "Detected and using 8 cores for processing\n",
      "2 replicates provided\n",
      "\n",
      "***Step 1: Generating sliding windows from the provided ChrAcc peaks file\n",
      "\n",
      "***Step 2: Assigning ATAC-STARR reads to sliding window bins\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.0\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m4 BAM files  \u001b[0m \u001b[0m                                   ||\n",
      "||                           \u001b[32mo\u001b[36m GM12878inGM12878_DNA_Rep1.filtered.n-sorte\u001b[0m ... \u001b[0m||\n",
      "||                           \u001b[32mo\u001b[36m GM12878inGM12878_DNA_Rep2.filtered.n-sorte\u001b[0m ... \u001b[0m||\n",
      "||                           \u001b[32mo\u001b[36m GM12878inGM12878_RNA_Rep1.filtered.n-sorte\u001b[0m ... \u001b[0m||\n",
      "||                           \u001b[32mo\u001b[36m GM12878inGM12878_RNA_Rep2.filtered.n-sorte\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcts.tsv\u001b[0m \u001b[0m                                         ||\n",
      "||                 Summary : \u001b[36mcts.tsv.summary\u001b[0m \u001b[0m                                 ||\n",
      "||              Annotation : \u001b[36mbins.saf (SAF)\u001b[0m \u001b[0m                                  ||\n",
      "||      Dir for temp files : \u001b[36m/data/hodges_lab/ATAC-STARR_V2/data/activity\u001b[0m ...\u001b[0m \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m8\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||              Paired-end : \u001b[36myes\u001b[0m \u001b[0m                                             ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mcounted\u001b[0m \u001b[0m                                         ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||          Chimeric reads : \u001b[36mcounted\u001b[0m \u001b[0m                                         ||\n",
      "||        Both ends mapped : \u001b[36mrequired\u001b[0m \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file bins.saf \u001b[0m... \u001b[0m                                         ||\n",
      "||    Features : \u001b[36m554\u001b[0m \u001b[0m                                                         ||\n",
      "||    Meta-features : \u001b[36m554\u001b[0m \u001b[0m                                                    ||\n",
      "||    Chromosomes/contigs : \u001b[36m1\u001b[0m \u001b[0m                                                ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file GM12878inGM12878_DNA_Rep1.filtered.n-sorted.bam... \u001b[0m       ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m30787318\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m3245 (0.0%)\u001b[0m \u001b[0m                         ||\n",
      "||    Running time : \u001b[36m0.52 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file GM12878inGM12878_DNA_Rep2.filtered.n-sorted.bam... \u001b[0m       ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m26516260\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m2780 (0.0%)\u001b[0m \u001b[0m                         ||\n",
      "||    Running time : \u001b[36m0.43 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam... \u001b[0m       ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m56277949\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m6161 (0.0%)\u001b[0m \u001b[0m                         ||\n",
      "||    Running time : \u001b[36m0.94 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file GM12878inGM12878_RNA_Rep2.filtered.n-sorted.bam... \u001b[0m       ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m67921622\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m7267 (0.0%)\u001b[0m \u001b[0m                         ||\n",
      "||    Running time : \u001b[36m1.13 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"/data/hodges_lab/ATAC-S \u001b[0m ||\n",
      "|| TARR_V2/data/activity/GM12878inGM12878/cts.tsv.summary\" \u001b[0m                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "***Step 3: Performing differential analysis on counted bins\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\n",
      "── Column specification ────────────────────────────────────────────────────────\n",
      "cols(\n",
      "  Bin_ID = col_character(),\n",
      "  Chr = col_character(),\n",
      "  Start = col_double(),\n",
      "  End = col_double(),\n",
      "  Strand = col_character(),\n",
      "  Length = col_double(),\n",
      "  DNA1 = col_double(),\n",
      "  DNA2 = col_double(),\n",
      "  RNA1 = col_double(),\n",
      "  RNA2 = col_double()\n",
      ")\n",
      "\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25hconverting counts to integer mode\n",
      "Warning message:\n",
      "In DESeqDataSet(se, design = design, ignoreRank) :\n",
      "  some variables in design formula are characters, converting to factors\n",
      "\u001b[?25hestimating size factors\n",
      "estimating dispersions\n",
      "gene-wise dispersion estimates: 8 workers\n",
      "mean-dispersion relationship\n",
      "final dispersion estimates, fitting model and testing: 8 workers\n",
      "\u001b[?25husing 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\n",
      "***Step 4: Merging active and silent bins into active/silent regulatory regions\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done!\n",
      "(jupyter) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "PEAKS='/data/hodges_lab/ATAC-STARR_V2/data/accessibility/genrich_ATAC-like_peaks/GM12878inGM12878_DNA_genrich_Test-set.narrowPeak'\n",
    "BAMS='/data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted'\n",
    "OUT_DIR='/data/hodges_lab/ATAC-STARR_V2/data/activity/GM12878inGM12878'\n",
    "\n",
    "#test on limited peak file\n",
    "python3 call_ATAC-STARR_regulatory-regions.py -i $PEAKS -o $OUT_DIR -q 0.1 \\\n",
    "    -d ${BAMS}/GM12878inGM12878_DNA_Rep1.filtered.n-sorted.bam ${BAMS}/GM12878inGM12878_DNA_Rep2.filtered.n-sorted.bam \\\n",
    "    -r ${BAMS}/GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam ${BAMS}/GM12878inGM12878_RNA_Rep2.filtered.n-sorted.bam"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "880206a5",
   "metadata": {},
   "source": [
    "## Test errors within the script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "de413f7b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(jupyter) (jupyter) (jupyter) PARAMETERS:\n",
      "ChrAcc Peaks File: /data/hodges_lab/ATAC-STARR_V2/data/accessibility/genrich_ATAC-like_peaks/GM12878inGM12878_DNA_genrich_Test-set.narrowPeak\n",
      "DNA Files: /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_DNA_Rep1.filtered.n-sorted.bam /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_DNA_Rep2.filtered.n-sorted.bam\n",
      "RNA Files: /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_RNA_Rep2.filtered.n-sorted.bam\n",
      "Output Directory: /data/hodges_lab/ATAC-STARR_V2/data/activity/GM12878inGM12878\n",
      "False-Discovery Rate: 0.1\n",
      "Detected and using 8 cores for processing\n",
      "ERROR: The number of DNA replicates provided does not match the number of RNA replicates provided\n",
      "(jupyter) (jupyter) (jupyter) PARAMETERS:\n",
      "ChrAcc Peaks File: /data/hodges_lab/ATAC-STARR_V2/data/accessibility/genrich_ATAC-like_peaks/GM12878inGM12878_DNA_genrich_Test-set.narrowPeak\n",
      "DNA Files: /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_DNA_Rep1.filtered.n-sorted.bam\n",
      "RNA Files: /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam\n",
      "Output Directory: /data/hodges_lab/ATAC-STARR_V2/data/activity/GM12878inGM12878\n",
      "False-Discovery Rate: 0.1\n",
      "Detected and using 8 cores for processing\n",
      "ERROR: Only one replicate detected. At least two replicates are required.\n",
      "(jupyter) (jupyter) (jupyter) PARAMETERS:\n",
      "ChrAcc Peaks File: /data/hodges_lab/ATAC-STARR_V2/data/accessibility/genrich_ATAC-like_peaks/GM12878inGM12878_DNA_genrich_Test-set.narrowPeak\n",
      "DNA Files: /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_DNA_Rep1.filtered.n-sorted.bam /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_DNA_Rep2.filtered.n-sorted.bam /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_DNA_Rep1.filtered.n-sorted.bam /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_DNA_Rep2.filtered.n-sorted.bam /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_DNA_Rep1.filtered.n-sorted.bam /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_DNA_Rep2.filtered.n-sorted.bam\n",
      "RNA Files: /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam\n",
      "Output Directory: /data/hodges_lab/ATAC-STARR_V2/data/activity/GM12878inGM12878\n",
      "False-Discovery Rate: 0.1\n",
      "Detected and using 8 cores for processing\n",
      "ERROR: More than five replicates detected. This script only allows up to five replicates in its current form.\n",
      "(jupyter) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "#test errors. \n",
    "\n",
    "#DNA and RNA differnt rep number\n",
    "python3 call_ATAC-STARR_regulatory-regions.py -i $PEAKS -o $OUT_DIR -q 0.1 \\\n",
    "    -d ${BAMS}/GM12878inGM12878_DNA_Rep1.filtered.n-sorted.bam ${BAMS}/GM12878inGM12878_DNA_Rep2.filtered.n-sorted.bam \\\n",
    "    -r ${BAMS}/GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam ${BAMS}/GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam ${BAMS}/GM12878inGM12878_RNA_Rep2.filtered.n-sorted.bam\n",
    "\n",
    "#Only one replicate\n",
    "python3 call_ATAC-STARR_regulatory-regions.py -i $PEAKS -o $OUT_DIR -q 0.1 \\\n",
    "    -d ${BAMS}/GM12878inGM12878_DNA_Rep1.filtered.n-sorted.bam \\\n",
    "    -r ${BAMS}/GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam\n",
    "\n",
    "#more than 5 replicates\n",
    "python3 call_ATAC-STARR_regulatory-regions.py -i $PEAKS -o $OUT_DIR -q 0.1 \\\n",
    "    -d ${BAMS}/GM12878inGM12878_DNA_Rep1.filtered.n-sorted.bam ${BAMS}/GM12878inGM12878_DNA_Rep2.filtered.n-sorted.bam \\\n",
    "    ${BAMS}/GM12878inGM12878_DNA_Rep1.filtered.n-sorted.bam ${BAMS}/GM12878inGM12878_DNA_Rep2.filtered.n-sorted.bam \\\n",
    "    ${BAMS}/GM12878inGM12878_DNA_Rep1.filtered.n-sorted.bam ${BAMS}/GM12878inGM12878_DNA_Rep2.filtered.n-sorted.bam \\\n",
    "    -r ${BAMS}/GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam ${BAMS}/GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam \\\n",
    "    ${BAMS}/GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam ${BAMS}/GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam \\\n",
    "    ${BAMS}/GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam ${BAMS}/GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5cb82c95",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(jupyter) mkdir: cannot create directory ‘tmp’: File exists\n",
      "(jupyter) (jupyter) (jupyter) (jupyter) ERROR: RNA-to-DNA_differential-analysis.r script cannot be located. Make sure it is contained in the same directory as this python script!\n",
      "(jupyter) (jupyter) (jupyter) (jupyter) (jupyter) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "#Test R script not being in the same directory error. \n",
    "mkdir tmp \n",
    "cp call_ATAC-STARR_regulatory-regions.py tmp/tmp.py\n",
    "cd tmp \n",
    "\n",
    "python3 tmp.py -i $PEAKS -o $OUT_DIR -q 0.1 \\\n",
    "    -d ${BAMS}/GM12878inGM12878_DNA_Rep1.filtered.n-sorted.bam ${BAMS}/GM12878inGM12878_DNA_Rep2.filtered.n-sorted.bam \\\n",
    "    -r ${BAMS}/GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam ${BAMS}/GM12878inGM12878_RNA_Rep2.filtered.n-sorted.bam\n",
    "\n",
    "rm tmp.py\n",
    "cd ..\n",
    "rmdir tmp"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3c7e0fd",
   "metadata": {},
   "source": [
    "## Test if script can recapture the same regions as before"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3ee191f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(jupyter) (jupyter) (jupyter) (jupyter) (jupyter) (jupyter) PARAMETERS:\n",
      "ChrAcc Peaks File: /data/hodges_lab/ATAC-STARR_V2/data/accessibility/genrich_ATAC-like_peaks/GM12878inGM12878_DNA_genrich_0.0001-qvalue.narrowPeak\n",
      "DNA Files: /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_DNA_Rep1.filtered.n-sorted.bam /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_DNA_Rep2.filtered.n-sorted.bam\n",
      "RNA Files: /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_RNA_Rep2.filtered.n-sorted.bam\n",
      "Output Directory: /data/hodges_lab/ATAC-STARR_V2/data/activity/GM12878inGM12878\n",
      "False-Discovery Rate: 0.1\n",
      "Detected and using 8 cores for processing\n",
      "2 replicates provided\n",
      "\n",
      "***Step 1: Generating sliding windows from the provided ChrAcc peaks file\n",
      "\n",
      "***Step 2: Assigning ATAC-STARR reads to sliding window bins\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.0\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m4 BAM files  \u001b[0m \u001b[0m                                   ||\n",
      "||                           \u001b[32mo\u001b[36m GM12878inGM12878_DNA_Rep1.filtered.n-sorte\u001b[0m ... \u001b[0m||\n",
      "||                           \u001b[32mo\u001b[36m GM12878inGM12878_DNA_Rep2.filtered.n-sorte\u001b[0m ... \u001b[0m||\n",
      "||                           \u001b[32mo\u001b[36m GM12878inGM12878_RNA_Rep1.filtered.n-sorte\u001b[0m ... \u001b[0m||\n",
      "||                           \u001b[32mo\u001b[36m GM12878inGM12878_RNA_Rep2.filtered.n-sorte\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcts.tsv\u001b[0m \u001b[0m                                         ||\n",
      "||                 Summary : \u001b[36mcts.tsv.summary\u001b[0m \u001b[0m                                 ||\n",
      "||              Annotation : \u001b[36mbins.saf (SAF)\u001b[0m \u001b[0m                                  ||\n",
      "||      Dir for temp files : \u001b[36m/data/hodges_lab/ATAC-STARR_V2/data/activity\u001b[0m ...\u001b[0m \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m8\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||              Paired-end : \u001b[36myes\u001b[0m \u001b[0m                                             ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mcounted\u001b[0m \u001b[0m                                         ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||          Chimeric reads : \u001b[36mcounted\u001b[0m \u001b[0m                                         ||\n",
      "||        Both ends mapped : \u001b[36mrequired\u001b[0m \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file bins.saf \u001b[0m... \u001b[0m                                         ||\n",
      "||    Features : \u001b[36m4540958\u001b[0m \u001b[0m                                                     ||\n",
      "||    Meta-features : \u001b[36m4540958\u001b[0m \u001b[0m                                                ||\n",
      "||    Chromosomes/contigs : \u001b[36m23\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file GM12878inGM12878_DNA_Rep1.filtered.n-sorted.bam... \u001b[0m       ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m30787318\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m22421492 (72.8%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m1.22 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file GM12878inGM12878_DNA_Rep2.filtered.n-sorted.bam... \u001b[0m       ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m26516260\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m19444815 (73.3%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m1.05 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam... \u001b[0m       ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m56277949\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m39550773 (70.3%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m1.29 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file GM12878inGM12878_RNA_Rep2.filtered.n-sorted.bam... \u001b[0m       ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m67921622\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m47752753 (70.3%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m2.35 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"/data/hodges_lab/ATAC-S \u001b[0m ||\n",
      "|| TARR_V2/data/activity/GM12878inGM12878/cts.tsv.summary\" \u001b[0m                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "***Step 3: Performing differential analysis on counted bins\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\n",
      "── Column specification ────────────────────────────────────────────────────────\n",
      "cols(\n",
      "  Bin_ID = col_character(),\n",
      "  Chr = col_character(),\n",
      "  Start = col_double(),\n",
      "  End = col_double(),\n",
      "  Strand = col_character(),\n",
      "  Length = col_double(),\n",
      "  DNA1 = col_double(),\n",
      "  DNA2 = col_double(),\n",
      "  RNA1 = col_double(),\n",
      "  RNA2 = col_double()\n",
      ")\n",
      "\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25hconverting counts to integer mode\n",
      "Warning message:\n",
      "In DESeqDataSet(se, design = design, ignoreRank) :\n",
      "  some variables in design formula are characters, converting to factors\n",
      "\u001b[?25hestimating size factors\n",
      "estimating dispersions\n",
      "gene-wise dispersion estimates: 8 workers\n",
      "mean-dispersion relationship\n",
      "final dispersion estimates, fitting model and testing: 8 workers\n",
      "\u001b[?25husing 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\n",
      "***Step 4: Merging active and silent bins into active/silent regulatory regions\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done!\n",
      "(jupyter) (jupyter) 15206 /data/hodges_lab/ATAC-STARR_V2/data/activity/GM12878inGM12878/active_regions.bed\n",
      "(jupyter) 20418 /data/hodges_lab/ATAC-STARR_V2/data/activity/GM12878inGM12878/silent_regions.bed\n",
      "(jupyter) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "#Test script to see if it captures the same number of active and silent regions as before.\n",
    "\n",
    "PEAKS='/data/hodges_lab/ATAC-STARR_V2/data/accessibility/genrich_ATAC-like_peaks/GM12878inGM12878_DNA_genrich_0.0001-qvalue.narrowPeak'\n",
    "BAMS='/data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted'\n",
    "OUT_DIR='/data/hodges_lab/ATAC-STARR_V2/data/activity/GM12878inGM12878'\n",
    "\n",
    "python3 call_ATAC-STARR_regulatory-regions.py -i $PEAKS -o $OUT_DIR -q 0.1 \\\n",
    "    -d ${BAMS}/GM12878inGM12878_DNA_Rep1.filtered.n-sorted.bam ${BAMS}/GM12878inGM12878_DNA_Rep2.filtered.n-sorted.bam \\\n",
    "    -r ${BAMS}/GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam ${BAMS}/GM12878inGM12878_RNA_Rep2.filtered.n-sorted.bam\n",
    "\n",
    "wc -l ${OUT_DIR}/active_regions.bed\n",
    "wc -l ${OUT_DIR}/silent_regions.bed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6451fac6",
   "metadata": {},
   "source": [
    "### Determine why there are more regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2a68d772",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(jupyter) 4540959 /data/hodges_lab/ATAC-STARR_V2/data/activity/GM12878inGM12878/bins.saf\n",
      "(jupyter) 4540959 /data/hodges_lab/ATAC-STARR_V2/data/activity/individual/bins/GM12878inGM12878_50bp-sliding-bins_10bp-step.saf\n",
      "(jupyter) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "#Do bin counts match? Yes. \n",
    "wc -l ${OUT_DIR}/bins.saf\n",
    "wc -l /data/hodges_lab/ATAC-STARR_V2/data/activity/individual/bins/GM12878inGM12878_50bp-sliding-bins_10bp-step.saf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "feb4fe82",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(jupyter) 123143 /data/hodges_lab/ATAC-STARR_V2/data/activity/GM12878inGM12878/active_bins.bed\n",
      "(jupyter) 123424 /data/hodges_lab/ATAC-STARR_V2/data/activity/individual/DESeq2_results/bin-method/GM12878inGM12878_active_0.1padj_50-bin.bed\n",
      "(jupyter) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "#Do active bins match? No. \n",
    "wc -l ${OUT_DIR}/active_bins.bed\n",
    "wc -l /data/hodges_lab/ATAC-STARR_V2/data/activity/individual/DESeq2_results/bin-method/GM12878inGM12878_active_0.1padj_50-bin.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1bb703e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(jupyter) 4540960 /data/hodges_lab/ATAC-STARR_V2/data/activity/GM12878inGM12878/cts.tsv\n",
      "(jupyter) 4540960 /data/hodges_lab/ATAC-STARR_V2/data/activity/individual/cts_matricies/bin-method/GM12878inGM12878_counts_bin-method_50bp-bins_with-duplicates.tsv\n",
      "(jupyter) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "#Does the cts matrix match? Yes. \n",
    "wc -l ${OUT_DIR}/cts.tsv\n",
    "wc -l /data/hodges_lab/ATAC-STARR_V2/data/activity/individual/cts_matricies/bin-method/GM12878inGM12878_counts_bin-method_50bp-bins_with-duplicates.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9bdf81af",
   "metadata": {},
   "source": [
    "It looks like the difference is at the DESeq2 step. I think it maybe because I changed the FDR from a > to a >=. \n",
    "\n",
    "I did change this operation, but only for active regions. The silent regions was still > and the number we get above is more than I got previously. \n",
    "\n",
    "Still though, lets check. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5fcd8f1",
   "metadata": {},
   "source": [
    "### Determine if the issue is at the less than/less than or equal to FDR cutoff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "957827c2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(jupyter) (jupyter) (jupyter) (jupyter) (jupyter) (jupyter) (jupyter) PARAMETERS:\n",
      "ChrAcc Peaks File: /data/hodges_lab/ATAC-STARR_V2/data/accessibility/genrich_ATAC-like_peaks/GM12878inGM12878_DNA_genrich_0.0001-qvalue.narrowPeak\n",
      "DNA Files: /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_DNA_Rep1.filtered.n-sorted.bam /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_DNA_Rep2.filtered.n-sorted.bam\n",
      "RNA Files: /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_RNA_Rep2.filtered.n-sorted.bam\n",
      "Output Directory: /data/hodges_lab/ATAC-STARR_V2/data/activity/GM12878inGM12878\n",
      "False-Discovery Rate: 0.1\n",
      "Detected and using 8 cores for processing\n",
      "2 replicates provided\n",
      "\n",
      "***Step 1: Generating sliding windows from the provided ChrAcc peaks file\n",
      "\n",
      "***Step 2: Assigning ATAC-STARR reads to sliding window bins\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.0\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m4 BAM files  \u001b[0m \u001b[0m                                   ||\n",
      "||                           \u001b[32mo\u001b[36m GM12878inGM12878_DNA_Rep1.filtered.n-sorte\u001b[0m ... \u001b[0m||\n",
      "||                           \u001b[32mo\u001b[36m GM12878inGM12878_DNA_Rep2.filtered.n-sorte\u001b[0m ... \u001b[0m||\n",
      "||                           \u001b[32mo\u001b[36m GM12878inGM12878_RNA_Rep1.filtered.n-sorte\u001b[0m ... \u001b[0m||\n",
      "||                           \u001b[32mo\u001b[36m GM12878inGM12878_RNA_Rep2.filtered.n-sorte\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcts.tsv\u001b[0m \u001b[0m                                         ||\n",
      "||                 Summary : \u001b[36mcts.tsv.summary\u001b[0m \u001b[0m                                 ||\n",
      "||              Annotation : \u001b[36mbins.saf (SAF)\u001b[0m \u001b[0m                                  ||\n",
      "||      Dir for temp files : \u001b[36m/data/hodges_lab/ATAC-STARR_V2/data/activity\u001b[0m ...\u001b[0m \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m8\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||              Paired-end : \u001b[36myes\u001b[0m \u001b[0m                                             ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mcounted\u001b[0m \u001b[0m                                         ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||          Chimeric reads : \u001b[36mcounted\u001b[0m \u001b[0m                                         ||\n",
      "||        Both ends mapped : \u001b[36mrequired\u001b[0m \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file bins.saf \u001b[0m... \u001b[0m                                         ||\n",
      "||    Features : \u001b[36m4540958\u001b[0m \u001b[0m                                                     ||\n",
      "||    Meta-features : \u001b[36m4540958\u001b[0m \u001b[0m                                                ||\n",
      "||    Chromosomes/contigs : \u001b[36m23\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file GM12878inGM12878_DNA_Rep1.filtered.n-sorted.bam... \u001b[0m       ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m30787318\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m22421492 (72.8%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m0.55 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file GM12878inGM12878_DNA_Rep2.filtered.n-sorted.bam... \u001b[0m       ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m26516260\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m19444815 (73.3%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m0.46 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam... \u001b[0m       ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m56277949\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m39550773 (70.3%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m0.95 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file GM12878inGM12878_RNA_Rep2.filtered.n-sorted.bam... \u001b[0m       ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m67921622\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m47752753 (70.3%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m1.17 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"/data/hodges_lab/ATAC-S \u001b[0m ||\n",
      "|| TARR_V2/data/activity/GM12878inGM12878/cts.tsv.summary\" \u001b[0m                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "***Step 3: Performing differential analysis on counted bins\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\n",
      "── Column specification ────────────────────────────────────────────────────────\n",
      "cols(\n",
      "  Bin_ID = col_character(),\n",
      "  Chr = col_character(),\n",
      "  Start = col_double(),\n",
      "  End = col_double(),\n",
      "  Strand = col_character(),\n",
      "  Length = col_double(),\n",
      "  DNA1 = col_double(),\n",
      "  DNA2 = col_double(),\n",
      "  RNA1 = col_double(),\n",
      "  RNA2 = col_double()\n",
      ")\n",
      "\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25hconverting counts to integer mode\n",
      "Warning message:\n",
      "In DESeqDataSet(se, design = design, ignoreRank) :\n",
      "  some variables in design formula are characters, converting to factors\n",
      "\u001b[?25hestimating size factors\n",
      "estimating dispersions\n",
      "gene-wise dispersion estimates: 8 workers\n",
      "mean-dispersion relationship\n",
      "final dispersion estimates, fitting model and testing: 8 workers\n",
      "\u001b[?25husing 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\n",
      "***Step 4: Merging active and silent bins into active/silent regulatory regions\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done!\n",
      "(jupyter) (jupyter) 15206 /data/hodges_lab/ATAC-STARR_V2/data/activity/GM12878inGM12878/active_regions.bed\n",
      "(jupyter) 20418 /data/hodges_lab/ATAC-STARR_V2/data/activity/GM12878inGM12878/silent_regions.bed\n",
      "(jupyter) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# I made a new R script that has this changed and I made a new python script with the updated R script. \n",
    "#Test script to see if it captures the same number of active and silent regions as before. Start: 8:40ish\n",
    "\n",
    "PEAKS='/data/hodges_lab/ATAC-STARR_V2/data/accessibility/genrich_ATAC-like_peaks/GM12878inGM12878_DNA_genrich_0.0001-qvalue.narrowPeak'\n",
    "BAMS='/data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted'\n",
    "OUT_DIR='/data/hodges_lab/ATAC-STARR_V2/data/activity/GM12878inGM12878'\n",
    "\n",
    "python3 call_ATAC-STARR_regulatory-regions_less-than-version.py -i $PEAKS -o $OUT_DIR -q 0.1 \\\n",
    "    -d ${BAMS}/GM12878inGM12878_DNA_Rep1.filtered.n-sorted.bam ${BAMS}/GM12878inGM12878_DNA_Rep2.filtered.n-sorted.bam \\\n",
    "    -r ${BAMS}/GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam ${BAMS}/GM12878inGM12878_RNA_Rep2.filtered.n-sorted.bam\n",
    "\n",
    "wc -l ${OUT_DIR}/active_regions.bed\n",
    "wc -l ${OUT_DIR}/silent_regions.bed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da71770b",
   "metadata": {},
   "source": [
    "It looks like this wasn't the issue. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d7bd6e90",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(jupyter) # Program:featureCounts v2.0.0; Command:\"featureCounts\" \"-p\" \"-O\" \"-B\" \"--minOverlap\" \"1\" \"-T\" \"8\" \"-F\" \"SAF\" \"-a\" \"/data/hodges_lab/ATAC-STARR_V2/data/activity/GM12878inGM12878/bins.saf\" \"-o\" \"/data/hodges_lab/ATAC-STARR_V2/data/activity/GM12878inGM12878/cts.tsv\" \"/data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_DNA_Rep1.filtered.n-sorted.bam\" \"/data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_DNA_Rep2.filtered.n-sorted.bam\" \"/data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam\" \"/data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_RNA_Rep2.filtered.n-sorted.bam\" \n",
      "Geneid\tChr\tStart\tEnd\tStrand\tLength\t/data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_DNA_Rep1.filtered.n-sorted.bam\t/data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_DNA_Rep2.filtered.n-sorted.bam\t/data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam\t/data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_RNA_Rep2.filtered.n-sorted.bam\n",
      "bin_1\tchr1\t778323\t778372\t.\t50\t11\t11\t40\t15\n",
      "bin_2\tchr1\t778333\t778382\t.\t50\t20\t19\t71\t52\n",
      "bin_3\tchr1\t778343\t778392\t.\t50\t26\t25\t109\t65\n",
      "bin_4\tchr1\t778353\t778402\t.\t50\t37\t39\t188\t149\n",
      "bin_5\tchr1\t778363\t778412\t.\t50\t47\t48\t212\t180\n",
      "bin_6\tchr1\t778373\t778422\t.\t50\t60\t53\t228\t223\n",
      "bin_7\tchr1\t778383\t778432\t.\t50\t61\t59\t245\t229\n",
      "bin_8\tchr1\t778393\t778442\t.\t50\t68\t66\t257\t229\n",
      "(jupyter) # Program:featureCounts v2.0.0; Command:\"featureCounts\" \"-p\" \"-O\" \"-B\" \"--minOverlap\" \"1\" \"-T\" \"12\" \"-F\" \"SAF\" \"-G\" \"/data/hodges_lab/hg38_genome/hg38.fa\" \"-a\" \"/data/hodges_lab/ATAC-STARR_V2/data/activity/individual/bins/GM12878inGM12878_50bp-sliding-bins_10bp-step.saf\" \"-o\" \"/data/hodges_lab/ATAC-STARR_V2/data/activity/individual/cts_matricies/bin-method/GM12878inGM12878_counts_bin-method_50bp-bins_with-duplicates.tsv\" \"/data/hodges_lab/ATAC-STARR_V2/data/bams/pos-sorted/GM12878inGM12878_DNA_Rep1.filtered.pos-sorted.bam\" \"/data/hodges_lab/ATAC-STARR_V2/data/bams/pos-sorted/GM12878inGM12878_DNA_Rep2.filtered.pos-sorted.bam\" \"/data/hodges_lab/ATAC-STARR_V2/data/bams/pos-sorted/GM12878inGM12878_RNA_Rep1.filtered.pos-sorted.bam\" \"/data/hodges_lab/ATAC-STARR_V2/data/bams/pos-sorted/GM12878inGM12878_RNA_Rep2.filtered.pos-sorted.bam\" \n",
      "Geneid\tChr\tStart\tEnd\tStrand\tLength\t/data/hodges_lab/ATAC-STARR_V2/data/bams/pos-sorted/GM12878inGM12878_DNA_Rep1.filtered.pos-sorted.bam\t/data/hodges_lab/ATAC-STARR_V2/data/bams/pos-sorted/GM12878inGM12878_DNA_Rep2.filtered.pos-sorted.bam\t/data/hodges_lab/ATAC-STARR_V2/data/bams/pos-sorted/GM12878inGM12878_RNA_Rep1.filtered.pos-sorted.bam\t/data/hodges_lab/ATAC-STARR_V2/data/bams/pos-sorted/GM12878inGM12878_RNA_Rep2.filtered.pos-sorted.bam\n",
      "bin_1\tchr1\t778322\t778372\t.\t51\t11\t11\t40\t15\n",
      "bin_2\tchr1\t778332\t778382\t.\t51\t20\t19\t71\t52\n",
      "bin_3\tchr1\t778342\t778392\t.\t51\t26\t25\t109\t65\n",
      "bin_4\tchr1\t778352\t778402\t.\t51\t37\t39\t188\t149\n",
      "bin_5\tchr1\t778362\t778412\t.\t51\t47\t48\t212\t180\n",
      "bin_6\tchr1\t778372\t778422\t.\t51\t60\t53\t228\t223\n",
      "bin_7\tchr1\t778382\t778432\t.\t51\t61\t59\t245\t229\n",
      "bin_8\tchr1\t778392\t778442\t.\t51\t68\t66\t257\t229\n",
      "(jupyter) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "#check input files:\n",
    "head ${OUT_DIR}/cts.tsv\n",
    "head /data/hodges_lab/ATAC-STARR_V2/data/activity/individual/cts_matricies/bin-method/GM12878inGM12878_counts_bin-method_50bp-bins_with-duplicates.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fb09458",
   "metadata": {},
   "source": [
    "It looks like the difference could be the featureCounts input being 1-based vs 0-based. \n",
    "\n",
    "Check it out. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51391ad7",
   "metadata": {},
   "source": [
    "### Determine if the problem is the 1 or 0 based difference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "60ad9702",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(jupyter) (jupyter) (jupyter) (jupyter) PARAMETERS:\n",
      "ChrAcc Peaks File: /data/hodges_lab/ATAC-STARR_V2/data/accessibility/genrich_ATAC-like_peaks/GM12878inGM12878_DNA_genrich_0.0001-qvalue.narrowPeak\n",
      "DNA Files: /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_DNA_Rep1.filtered.n-sorted.bam /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_DNA_Rep2.filtered.n-sorted.bam\n",
      "RNA Files: /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_RNA_Rep2.filtered.n-sorted.bam\n",
      "Output Directory: /data/hodges_lab/ATAC-STARR_V2/data/activity/GM12878inGM12878\n",
      "False-Discovery Rate: 0.1\n",
      "Detected and using 8 cores for processing\n",
      "2 replicates provided\n",
      "\n",
      "***Step 1: Generating sliding windows from the provided ChrAcc peaks file\n",
      "\n",
      "***Step 2: Assigning ATAC-STARR reads to sliding window bins\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.0\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m4 BAM files  \u001b[0m \u001b[0m                                   ||\n",
      "||                           \u001b[32mo\u001b[36m GM12878inGM12878_DNA_Rep1.filtered.n-sorte\u001b[0m ... \u001b[0m||\n",
      "||                           \u001b[32mo\u001b[36m GM12878inGM12878_DNA_Rep2.filtered.n-sorte\u001b[0m ... \u001b[0m||\n",
      "||                           \u001b[32mo\u001b[36m GM12878inGM12878_RNA_Rep1.filtered.n-sorte\u001b[0m ... \u001b[0m||\n",
      "||                           \u001b[32mo\u001b[36m GM12878inGM12878_RNA_Rep2.filtered.n-sorte\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcts.tsv\u001b[0m \u001b[0m                                         ||\n",
      "||                 Summary : \u001b[36mcts.tsv.summary\u001b[0m \u001b[0m                                 ||\n",
      "||              Annotation : \u001b[36mbins.saf (SAF)\u001b[0m \u001b[0m                                  ||\n",
      "||      Dir for temp files : \u001b[36m/data/hodges_lab/ATAC-STARR_V2/data/activity\u001b[0m ...\u001b[0m \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m8\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||              Paired-end : \u001b[36myes\u001b[0m \u001b[0m                                             ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mcounted\u001b[0m \u001b[0m                                         ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||          Chimeric reads : \u001b[36mcounted\u001b[0m \u001b[0m                                         ||\n",
      "||        Both ends mapped : \u001b[36mrequired\u001b[0m \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file bins.saf \u001b[0m... \u001b[0m                                         ||\n",
      "||    Features : \u001b[36m4540958\u001b[0m \u001b[0m                                                     ||\n",
      "||    Meta-features : \u001b[36m4540958\u001b[0m \u001b[0m                                                ||\n",
      "||    Chromosomes/contigs : \u001b[36m23\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file GM12878inGM12878_DNA_Rep1.filtered.n-sorted.bam... \u001b[0m       ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m30787318\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m22422231 (72.8%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m0.63 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file GM12878inGM12878_DNA_Rep2.filtered.n-sorted.bam... \u001b[0m       ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m26516260\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m19445438 (73.3%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m0.53 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam... \u001b[0m       ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m56277949\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m39552311 (70.3%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m1.23 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file GM12878inGM12878_RNA_Rep2.filtered.n-sorted.bam... \u001b[0m       ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m67921622\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m47754483 (70.3%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m1.38 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"/data/hodges_lab/ATAC-S \u001b[0m ||\n",
      "|| TARR_V2/data/activity/GM12878inGM12878/cts.tsv.summary\" \u001b[0m                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "***Step 3: Performing differential analysis on counted bins\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\n",
      "── Column specification ────────────────────────────────────────────────────────\n",
      "cols(\n",
      "  Bin_ID = col_character(),\n",
      "  Chr = col_character(),\n",
      "  Start = col_double(),\n",
      "  End = col_double(),\n",
      "  Strand = col_character(),\n",
      "  Length = col_double(),\n",
      "  DNA1 = col_double(),\n",
      "  DNA2 = col_double(),\n",
      "  RNA1 = col_double(),\n",
      "  RNA2 = col_double()\n",
      ")\n",
      "\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25hconverting counts to integer mode\n",
      "Warning message:\n",
      "In DESeqDataSet(se, design = design, ignoreRank) :\n",
      "  some variables in design formula are characters, converting to factors\n",
      "\u001b[?25hestimating size factors\n",
      "estimating dispersions\n",
      "gene-wise dispersion estimates: 8 workers\n",
      "mean-dispersion relationship\n",
      "final dispersion estimates, fitting model and testing: 8 workers\n",
      "\u001b[?25husing 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\n",
      "***Step 4: Merging active and silent bins into active/silent regulatory regions\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done!\n",
      "(jupyter) (jupyter) 14617 /data/hodges_lab/ATAC-STARR_V2/data/activity/GM12878inGM12878/active_regions.bed\n",
      "(jupyter) 19393 /data/hodges_lab/ATAC-STARR_V2/data/activity/GM12878inGM12878/silent_regions.bed\n",
      "(jupyter) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "PEAKS='/data/hodges_lab/ATAC-STARR_V2/data/accessibility/genrich_ATAC-like_peaks/GM12878inGM12878_DNA_genrich_0.0001-qvalue.narrowPeak'\n",
    "BAMS='/data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted'\n",
    "OUT_DIR='/data/hodges_lab/ATAC-STARR_V2/data/activity/GM12878inGM12878'\n",
    "\n",
    "python3 call_ATAC-STARR_regulatory-regions_0-based.py -i $PEAKS -o $OUT_DIR -q 0.1 \\\n",
    "    -d ${BAMS}/GM12878inGM12878_DNA_Rep1.filtered.n-sorted.bam ${BAMS}/GM12878inGM12878_DNA_Rep2.filtered.n-sorted.bam \\\n",
    "    -r ${BAMS}/GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam ${BAMS}/GM12878inGM12878_RNA_Rep2.filtered.n-sorted.bam\n",
    "\n",
    "wc -l ${OUT_DIR}/active_regions.bed\n",
    "wc -l ${OUT_DIR}/silent_regions.bed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03fcf83d",
   "metadata": {},
   "source": [
    "Yep! This was the difference. 1-based is what we should be using, so lets go with that. I also changed the R script to format .bed output files to 0-based since I hadn't done this yet. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "155e801c",
   "metadata": {},
   "source": [
    "## Re-run finalized script to get GG regulatory regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "fb48894f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(jupyter) (jupyter) (jupyter) (jupyter) PARAMETERS:\n",
      "ChrAcc Peaks File: /data/hodges_lab/ATAC-STARR_V2/data/accessibility/genrich_ATAC-like_peaks/GM12878inGM12878_DNA_genrich_0.0001-qvalue.narrowPeak\n",
      "DNA Files: /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_DNA_Rep1.filtered.n-sorted.bam /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_DNA_Rep2.filtered.n-sorted.bam\n",
      "RNA Files: /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_RNA_Rep2.filtered.n-sorted.bam\n",
      "Output Directory: /data/hodges_lab/ATAC-STARR_V2/data/activity/GM12878inGM12878\n",
      "False-Discovery Rate: 0.1\n",
      "Detected and using 8 cores for processing\n",
      "2 replicates provided\n",
      "\n",
      "***Step 1: Generating sliding windows from the provided ChrAcc peaks file\n",
      "\n",
      "***Step 2: Assigning ATAC-STARR reads to sliding window bins\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.0\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m4 BAM files  \u001b[0m \u001b[0m                                   ||\n",
      "||                           \u001b[32mo\u001b[36m GM12878inGM12878_DNA_Rep1.filtered.n-sorte\u001b[0m ... \u001b[0m||\n",
      "||                           \u001b[32mo\u001b[36m GM12878inGM12878_DNA_Rep2.filtered.n-sorte\u001b[0m ... \u001b[0m||\n",
      "||                           \u001b[32mo\u001b[36m GM12878inGM12878_RNA_Rep1.filtered.n-sorte\u001b[0m ... \u001b[0m||\n",
      "||                           \u001b[32mo\u001b[36m GM12878inGM12878_RNA_Rep2.filtered.n-sorte\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcts.tsv\u001b[0m \u001b[0m                                         ||\n",
      "||                 Summary : \u001b[36mcts.tsv.summary\u001b[0m \u001b[0m                                 ||\n",
      "||              Annotation : \u001b[36mbins.saf (SAF)\u001b[0m \u001b[0m                                  ||\n",
      "||      Dir for temp files : \u001b[36m/data/hodges_lab/ATAC-STARR_V2/data/activity\u001b[0m ...\u001b[0m \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m8\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||              Paired-end : \u001b[36myes\u001b[0m \u001b[0m                                             ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mcounted\u001b[0m \u001b[0m                                         ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||          Chimeric reads : \u001b[36mcounted\u001b[0m \u001b[0m                                         ||\n",
      "||        Both ends mapped : \u001b[36mrequired\u001b[0m \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file bins.saf \u001b[0m... \u001b[0m                                         ||\n",
      "||    Features : \u001b[36m4540958\u001b[0m \u001b[0m                                                     ||\n",
      "||    Meta-features : \u001b[36m4540958\u001b[0m \u001b[0m                                                ||\n",
      "||    Chromosomes/contigs : \u001b[36m23\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file GM12878inGM12878_DNA_Rep1.filtered.n-sorted.bam... \u001b[0m       ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m30787318\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m22421492 (72.8%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m0.67 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file GM12878inGM12878_DNA_Rep2.filtered.n-sorted.bam... \u001b[0m       ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m26516260\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m19444815 (73.3%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m0.65 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam... \u001b[0m       ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m56277949\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m39550773 (70.3%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m1.40 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file GM12878inGM12878_RNA_Rep2.filtered.n-sorted.bam... \u001b[0m       ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m67921622\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m47752753 (70.3%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m1.60 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"/data/hodges_lab/ATAC-S \u001b[0m ||\n",
      "|| TARR_V2/data/activity/GM12878inGM12878/cts.tsv.summary\" \u001b[0m                   ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "***Step 3: Performing differential analysis on counted bins\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\n",
      "── Column specification ────────────────────────────────────────────────────────\n",
      "cols(\n",
      "  Bin_ID = col_character(),\n",
      "  Chr = col_character(),\n",
      "  Start = col_double(),\n",
      "  End = col_double(),\n",
      "  Strand = col_character(),\n",
      "  Length = col_double(),\n",
      "  DNA1 = col_double(),\n",
      "  DNA2 = col_double(),\n",
      "  RNA1 = col_double(),\n",
      "  RNA2 = col_double()\n",
      ")\n",
      "\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25hconverting counts to integer mode\n",
      "Warning message:\n",
      "In DESeqDataSet(se, design = design, ignoreRank) :\n",
      "  some variables in design formula are characters, converting to factors\n",
      "\u001b[?25hestimating size factors\n",
      "estimating dispersions\n",
      "gene-wise dispersion estimates: 8 workers\n",
      "mean-dispersion relationship\n",
      "final dispersion estimates, fitting model and testing: 8 workers\n",
      "\u001b[?25husing 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\n",
      "***Step 4: Merging active and silent bins into active/silent regulatory regions\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done!\n",
      "(jupyter) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "PEAKS='/data/hodges_lab/ATAC-STARR_V2/data/accessibility/genrich_ATAC-like_peaks/GM12878inGM12878_DNA_genrich_0.0001-qvalue.narrowPeak'\n",
    "BAMS='/data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted'\n",
    "OUT_DIR='/data/hodges_lab/ATAC-STARR_V2/data/activity/GM12878inGM12878'\n",
    "\n",
    "python3 call_ATAC-STARR_regulatory-regions.py -i $PEAKS -o $OUT_DIR -q 0.1 \\\n",
    "    -d ${BAMS}/GM12878inGM12878_DNA_Rep1.filtered.n-sorted.bam ${BAMS}/GM12878inGM12878_DNA_Rep2.filtered.n-sorted.bam \\\n",
    "    -r ${BAMS}/GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam ${BAMS}/GM12878inGM12878_RNA_Rep2.filtered.n-sorted.bam\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a984f062",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(jupyter) (jupyter) (jupyter) (jupyter) (jupyter) (jupyter) (jupyter) PARAMETERS:\n",
      "ChrAcc Peaks File: /data/hodges_lab/ATAC-STARR_V2/data/accessibility/genrich_ATAC-like_peaks/GM12878inGM12878_DNA_genrich_0.0001-qvalue.narrowPeak\n",
      "DNA Files: /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_DNA_Rep1.filtered.n-sorted.bam /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_DNA_Rep2.filtered.n-sorted.bam\n",
      "RNA Files: /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam /data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted/GM12878inGM12878_RNA_Rep2.filtered.n-sorted.bam\n",
      "Output Directory: /data/hodges_lab/ATAC-STARR_V2/data/activity/GM12878inGM12878_FDR-0.05\n",
      "False-Discovery Rate: 0.05\n",
      "Detected and using 8 cores for processing\n",
      "2 replicates provided\n",
      "\n",
      "***Step 1: Generating sliding windows from the provided ChrAcc peaks file\n",
      "\n",
      "***Step 2: Assigning ATAC-STARR reads to sliding window bins\n",
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.0\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m4 BAM files  \u001b[0m \u001b[0m                                   ||\n",
      "||                           \u001b[32mo\u001b[36m GM12878inGM12878_DNA_Rep1.filtered.n-sorte\u001b[0m ... \u001b[0m||\n",
      "||                           \u001b[32mo\u001b[36m GM12878inGM12878_DNA_Rep2.filtered.n-sorte\u001b[0m ... \u001b[0m||\n",
      "||                           \u001b[32mo\u001b[36m GM12878inGM12878_RNA_Rep1.filtered.n-sorte\u001b[0m ... \u001b[0m||\n",
      "||                           \u001b[32mo\u001b[36m GM12878inGM12878_RNA_Rep2.filtered.n-sorte\u001b[0m ... \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mcts.tsv\u001b[0m \u001b[0m                                         ||\n",
      "||                 Summary : \u001b[36mcts.tsv.summary\u001b[0m \u001b[0m                                 ||\n",
      "||              Annotation : \u001b[36mbins.saf (SAF)\u001b[0m \u001b[0m                                  ||\n",
      "||      Dir for temp files : \u001b[36m/data/hodges_lab/ATAC-STARR_V2/data/activity\u001b[0m ...\u001b[0m \u001b[0m||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m8\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||              Paired-end : \u001b[36myes\u001b[0m \u001b[0m                                             ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mcounted\u001b[0m \u001b[0m                                         ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||          Chimeric reads : \u001b[36mcounted\u001b[0m \u001b[0m                                         ||\n",
      "||        Both ends mapped : \u001b[36mrequired\u001b[0m \u001b[0m                                        ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file bins.saf \u001b[0m... \u001b[0m                                         ||\n",
      "||    Features : \u001b[36m4540958\u001b[0m \u001b[0m                                                     ||\n",
      "||    Meta-features : \u001b[36m4540958\u001b[0m \u001b[0m                                                ||\n",
      "||    Chromosomes/contigs : \u001b[36m23\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file GM12878inGM12878_DNA_Rep1.filtered.n-sorted.bam... \u001b[0m       ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m30787318\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m22421492 (72.8%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m0.79 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file GM12878inGM12878_DNA_Rep2.filtered.n-sorted.bam... \u001b[0m       ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m26516260\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m19444815 (73.3%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m0.61 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam... \u001b[0m       ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m56277949\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m39550773 (70.3%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m1.29 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file GM12878inGM12878_RNA_Rep2.filtered.n-sorted.bam... \u001b[0m       ||\n",
      "||    Paired-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m67921622\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m47752753 (70.3%)\u001b[0m \u001b[0m                    ||\n",
      "||    Running time : \u001b[36m1.89 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"/data/hodges_lab/ATAC-S \u001b[0m ||\n",
      "|| TARR_V2/data/activity/GM12878inGM12878_FDR-0.05/cts.tsv.summary\" \u001b[0m          ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "\n",
      "***Step 3: Performing differential analysis on counted bins\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\n",
      "── Column specification ────────────────────────────────────────────────────────\n",
      "cols(\n",
      "  Bin_ID = col_character(),\n",
      "  Chr = col_character(),\n",
      "  Start = col_double(),\n",
      "  End = col_double(),\n",
      "  Strand = col_character(),\n",
      "  Length = col_double(),\n",
      "  DNA1 = col_double(),\n",
      "  DNA2 = col_double(),\n",
      "  RNA1 = col_double(),\n",
      "  RNA2 = col_double()\n",
      ")\n",
      "\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25hconverting counts to integer mode\n",
      "Warning message:\n",
      "In DESeqDataSet(se, design = design, ignoreRank) :\n",
      "  some variables in design formula are characters, converting to factors\n",
      "\u001b[?25hestimating size factors\n",
      "estimating dispersions\n",
      "gene-wise dispersion estimates: 8 workers\n",
      "mean-dispersion relationship\n",
      "final dispersion estimates, fitting model and testing: 8 workers\n",
      "\u001b[?25husing 'apeglm' for LFC shrinkage. If used in published research, please cite:\n",
      "    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for\n",
      "    sequence count data: removing the noise and preserving large differences.\n",
      "    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\n",
      "***Step 4: Merging active and silent bins into active/silent regulatory regions\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done!\n",
      "(jupyter) (jupyter) 9645 /data/hodges_lab/ATAC-STARR_V2/data/activity/GM12878inGM12878_FDR-0.05/active_regions.bed\n",
      "(jupyter) 16967 /data/hodges_lab/ATAC-STARR_V2/data/activity/GM12878inGM12878_FDR-0.05/silent_regions.bed\n",
      "(jupyter) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "#Try adjusting FDR\n",
    "mkdir /data/hodges_lab/ATAC-STARR_V2/data/activity/GM12878inGM12878_FDR-0.05\n",
    "\n",
    "PEAKS='/data/hodges_lab/ATAC-STARR_V2/data/accessibility/genrich_ATAC-like_peaks/GM12878inGM12878_DNA_genrich_0.0001-qvalue.narrowPeak'\n",
    "BAMS='/data/hodges_lab/ATAC-STARR_V2/data/bams/n-sorted'\n",
    "OUT_DIR='/data/hodges_lab/ATAC-STARR_V2/data/activity/GM12878inGM12878_FDR-0.05'\n",
    "\n",
    "python3 call_ATAC-STARR_regulatory-regions.py -i $PEAKS -o $OUT_DIR -q 0.05 \\\n",
    "    -d ${BAMS}/GM12878inGM12878_DNA_Rep1.filtered.n-sorted.bam ${BAMS}/GM12878inGM12878_DNA_Rep2.filtered.n-sorted.bam \\\n",
    "    -r ${BAMS}/GM12878inGM12878_RNA_Rep1.filtered.n-sorted.bam ${BAMS}/GM12878inGM12878_RNA_Rep2.filtered.n-sorted.bam\n",
    "\n",
    "wc -l ${OUT_DIR}/active_regions.bed\n",
    "wc -l ${OUT_DIR}/silent_regions.bed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0e7102c",
   "metadata": {},
   "source": [
    "# ATAC-STARR Fastq Processing Script"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "437a41a6",
   "metadata": {},
   "source": [
    "## Introduction\n",
    "\n",
    "I made a python script to process fastq files for ATAC-STARR data. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a545e3b",
   "metadata": {},
   "source": [
    "## Prepare Testing Environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "00320cba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘/scratch/hansetj1/script_testing’: File exists\n",
      "mkdir: cannot create directory ‘/scratch/hansetj1/script_testing/fastq_processing’: File exists\n",
      "mkdir: cannot create directory ‘/scratch/hansetj1/script_testing/fastq_processing/raw_reads’: File exists\n",
      "mkdir: cannot create directory ‘/scratch/hansetj1/script_testing/fastq_processing/bams’: File exists\n",
      "mkdir: cannot create directory ‘/scratch/hansetj1/script_testing/fastq_processing/qc’: File exists\n",
      "mkdir: cannot create directory ‘/scratch/hansetj1/script_testing/fastq_processing/trimmed_reads’: File exists\n"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "#Make test folders on scratch\n",
    "mkdir /scratch/hansetj1/script_testing\n",
    "mkdir /scratch/hansetj1/script_testing/fastq_processing\n",
    "mkdir /scratch/hansetj1/script_testing/fastq_processing/raw_reads\n",
    "mkdir /scratch/hansetj1/script_testing/fastq_processing/bams\n",
    "mkdir /scratch/hansetj1/script_testing/fastq_processing/qc\n",
    "mkdir /scratch/hansetj1/script_testing/fastq_processing/trimmed_reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "bc7f3131",
   "metadata": {},
   "outputs": [],
   "source": [
    "#move fastq files to new folder\n",
    "\n",
    "dir='/data/hodges_lab/ATAC-STARR/data/1_FASTQ'\n",
    "\n",
    "gunzip -c ${dir}/3108-TH-1-TAAGGCGA-GTAAGGAG_S401_R1_001.fastq.gz > /scratch/hansetj1/script_testing/fastq_processing/raw_reads/read1.fq\n",
    "gunzip -c ${dir}/3108-TH-1-TAAGGCGA-GTAAGGAG_S401_R2_001.fastq.gz > /scratch/hansetj1/script_testing/fastq_processing/raw_reads/read2.fq\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7a74ae3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Subsample fastq files for quick processing using seqtk\n",
    "\n",
    "dir='/scratch/hansetj1/script_testing/fastq_processing/raw_reads'\n",
    "\n",
    "# FASTQ R1\n",
    "seqtk sample -s 123 ${dir}/read1.fq 10000 > ${dir}/sub_read1.fq\n",
    " \n",
    "# FASTQ R2\n",
    "seqtk sample -s 123 ${dir}/read2.fq 10000 > ${dir}/sub_read2.fq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25ef3530",
   "metadata": {},
   "source": [
    "## Test help functionality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "53b738ba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: fastq_processing.py [-h] [-a INPUT_R1] [-b INPUT_R2] [-n NAME]\n",
      "                           [-e BLACKLIST_BED] [-t OUTPUT_TRIMMED]\n",
      "                           [-o OUTPUT_BAM] [-r OUTPUT_QC] [-g GENOME]\n",
      "\n",
      "process ATAC-STARR fastq files\n",
      "\n",
      "required arguments:\n",
      "  -a INPUT_R1, --input_R1 INPUT_R1\n",
      "                        raw fastq paired-end read 1 file\n",
      "  -b INPUT_R2, --input_R2 INPUT_R2\n",
      "                        raw fastq paired-end read 2 file\n",
      "  -n NAME, --name NAME  basename to assign to sample\n",
      "  -t OUTPUT_TRIMMED, --output_trimmed OUTPUT_TRIMMED\n",
      "                        directory to deposit trimmed reads\n",
      "  -o OUTPUT_BAM, --output_bam OUTPUT_BAM\n",
      "                        directory to deposit mapped reads\n",
      "  -r OUTPUT_QC, --output_qc OUTPUT_QC\n",
      "                        directory to deposit QC and readcount files\n",
      "  -g GENOME, --genome GENOME\n",
      "                        bowtie index of genome to map to\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "  -e BLACKLIST_BED, --blacklist_bed BLACKLIST_BED\n",
      "                        bedfile of ENCODE blacklist regions (default is none\n",
      "                        and skips this step)\n"
     ]
    }
   ],
   "source": [
    "python3 ./fastq_processing/fastq_processing.py --help"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42205a7c",
   "metadata": {},
   "source": [
    "## Test complete run-through with sub-sampled data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "b4e89628",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PARAMETERS:\n",
      "Fastq R1: /scratch/hansetj1/script_testing/fastq_processing/raw_reads/sub_read1.fq\n",
      "Fastq R2: /scratch/hansetj1/script_testing/fastq_processing/raw_reads/sub_read2.fq\n",
      "Sample name: test_set\n",
      "Blacklist Bedfile /home/hansetj1/hg38_encode_blacklist_ENCFF356LFX.bed\n",
      "Trimmed reads directory: /scratch/hansetj1/script_testing/fastq_processing/trimmed_reads\n",
      "Mapped reads directory: /scratch/hansetj1/script_testing/fastq_processing/bams\n",
      "QC results directory: /scratch/hansetj1/script_testing/fastq_processing/qc\n",
      "Genome: /data/hodges_lab/hg38_genome/bt2/hg38\n",
      "Detected and using 8 cores for processing\n",
      "\n",
      "***Trimming raw reads\n",
      "Multicore support not enabled. Proceeding with single-core trimming.\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 1.18\n",
      "single-core operation.\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output will be written into the directory: /gpfs23/scratch/hansetj1/script_testing/fastq_processing/trimmed_reads/\n",
      "Using user-specified basename (>>test_set<<) instead of deriving the filename from the input file(s)\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> /scratch/hansetj1/script_testing/fastq_processing/raw_reads/sub_read1.fq <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "Nextera\t4704\tCTGTCTCTTATA\t10000\t47.04\n",
      "smallRNA\t0\tTGGAATTCTCGG\t10000\t0.00\n",
      "Illumina\t0\tAGATCGGAAGAGC\t10000\t0.00\n",
      "Using Nextera adapter for trimming (count: 4704). Second best hit was smallRNA (count: 0)\n",
      "\n",
      "Writing report to '/gpfs23/scratch/hansetj1/script_testing/fastq_processing/trimmed_reads/sub_read1.fq_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /scratch/hansetj1/script_testing/fastq_processing/raw_reads/sub_read1.fq\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.6\n",
      "Cutadapt version: 1.18\n",
      "Number of cores used for trimming: 1\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'CTGTCTCTTATA' (Nextera Transposase sequence; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Running FastQC on the data once trimming has completed\n",
      "Running FastQC with the following extra arguments: '--outdir /scratch/hansetj1/script_testing/fastq_processing/qc'\n",
      "\n",
      "Cutadapt seems to be reasonably up-to-date. Setting -j 1\n",
      "Writing final adapter and quality trimmed output to sub_read1_trimmed.fq\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'CTGTCTCTTATA' from file /scratch/hansetj1/script_testing/fastq_processing/raw_reads/sub_read1.fq <<< \n",
      "This is cutadapt 1.18 with Python 3.6.13\n",
      "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a CTGTCTCTTATA /scratch/hansetj1/script_testing/fastq_processing/raw_reads/sub_read1.fq\n",
      "Processing reads on 1 core in single-end mode ...\n",
      "Finished in 0.14 s (14 us/read; 4.16 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:                  10,000\n",
      "Reads with adapters:                     7,375 (73.8%)\n",
      "Reads written (passing filters):        10,000 (100.0%)\n",
      "\n",
      "Total basepairs processed:     1,496,800 bp\n",
      "Quality-trimmed:                   8,932 bp (0.6%)\n",
      "Total written (filtered):      1,157,015 bp (77.3%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: CTGTCTCTTATA; Type: regular 3'; Length: 12; Trimmed: 7375 times.\n",
      "\n",
      "No. of allowed errors:\n",
      "0-9 bp: 0; 10-12 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 13.4%\n",
      "  C: 40.3%\n",
      "  G: 25.5%\n",
      "  T: 20.7%\n",
      "  none/other: 0.1%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t1710\t2500.0\t0\t1710\n",
      "2\t426\t625.0\t0\t426\n",
      "3\t162\t156.2\t0\t162\n",
      "4\t56\t39.1\t0\t56\n",
      "5\t34\t9.8\t0\t34\n",
      "6\t34\t2.4\t0\t34\n",
      "7\t24\t0.6\t0\t24\n",
      "8\t34\t0.2\t0\t34\n",
      "9\t46\t0.0\t0\t46\n",
      "11\t48\t0.0\t1\t47 1\n",
      "12\t1\t0.0\t1\t1\n",
      "13\t57\t0.0\t1\t57\n",
      "14\t1\t0.0\t1\t0 1\n",
      "15\t79\t0.0\t1\t77 2\n",
      "16\t3\t0.0\t1\t2 1\n",
      "17\t27\t0.0\t1\t27\n",
      "18\t28\t0.0\t1\t28\n",
      "19\t27\t0.0\t1\t25 2\n",
      "20\t39\t0.0\t1\t39\n",
      "21\t30\t0.0\t1\t29 1\n",
      "22\t85\t0.0\t1\t85\n",
      "24\t1\t0.0\t1\t0 1\n",
      "25\t43\t0.0\t1\t42 1\n",
      "26\t27\t0.0\t1\t26 1\n",
      "27\t80\t0.0\t1\t79 1\n",
      "29\t37\t0.0\t1\t35 2\n",
      "30\t129\t0.0\t1\t128 1\n",
      "31\t1\t0.0\t1\t0 1\n",
      "33\t1\t0.0\t1\t0 1\n",
      "34\t39\t0.0\t1\t39\n",
      "35\t67\t0.0\t1\t66 1\n",
      "36\t60\t0.0\t1\t60\n",
      "37\t2\t0.0\t1\t0 2\n",
      "38\t1\t0.0\t1\t1\n",
      "39\t27\t0.0\t1\t27\n",
      "40\t41\t0.0\t1\t38 3\n",
      "41\t89\t0.0\t1\t88 1\n",
      "42\t45\t0.0\t1\t44 1\n",
      "43\t6\t0.0\t1\t6\n",
      "44\t51\t0.0\t1\t49 2\n",
      "45\t110\t0.0\t1\t106 4\n",
      "46\t14\t0.0\t1\t14\n",
      "47\t45\t0.0\t1\t44 1\n",
      "48\t292\t0.0\t1\t289 3\n",
      "49\t55\t0.0\t1\t55\n",
      "50\t25\t0.0\t1\t23 2\n",
      "51\t168\t0.0\t1\t165 3\n",
      "52\t41\t0.0\t1\t41\n",
      "53\t5\t0.0\t1\t4 1\n",
      "54\t72\t0.0\t1\t72\n",
      "55\t64\t0.0\t1\t63 1\n",
      "56\t28\t0.0\t1\t28\n",
      "57\t21\t0.0\t1\t20 1\n",
      "58\t7\t0.0\t1\t7\n",
      "59\t46\t0.0\t1\t46\n",
      "60\t6\t0.0\t1\t6\n",
      "61\t13\t0.0\t1\t13\n",
      "62\t37\t0.0\t1\t36 1\n",
      "63\t12\t0.0\t1\t11 1\n",
      "64\t4\t0.0\t1\t4\n",
      "65\t43\t0.0\t1\t43\n",
      "66\t66\t0.0\t1\t66\n",
      "67\t14\t0.0\t1\t14\n",
      "68\t35\t0.0\t1\t33 2\n",
      "69\t50\t0.0\t1\t49 1\n",
      "70\t88\t0.0\t1\t87 1\n",
      "71\t39\t0.0\t1\t39\n",
      "72\t6\t0.0\t1\t6\n",
      "73\t8\t0.0\t1\t6 2\n",
      "74\t22\t0.0\t1\t21 1\n",
      "75\t41\t0.0\t1\t40 1\n",
      "76\t51\t0.0\t1\t51\n",
      "77\t50\t0.0\t1\t49 1\n",
      "78\t42\t0.0\t1\t42\n",
      "79\t47\t0.0\t1\t45 2\n",
      "80\t49\t0.0\t1\t48 1\n",
      "81\t51\t0.0\t1\t50 1\n",
      "82\t55\t0.0\t1\t54 1\n",
      "83\t58\t0.0\t1\t54 4\n",
      "84\t66\t0.0\t1\t65 1\n",
      "85\t48\t0.0\t1\t46 2\n",
      "86\t64\t0.0\t1\t60 4\n",
      "87\t51\t0.0\t1\t50 1\n",
      "88\t58\t0.0\t1\t57 1\n",
      "89\t60\t0.0\t1\t60\n",
      "90\t56\t0.0\t1\t56\n",
      "91\t57\t0.0\t1\t57\n",
      "92\t51\t0.0\t1\t51\n",
      "93\t39\t0.0\t1\t39\n",
      "94\t56\t0.0\t1\t54 2\n",
      "95\t58\t0.0\t1\t57 1\n",
      "96\t71\t0.0\t1\t69 2\n",
      "97\t68\t0.0\t1\t65 3\n",
      "98\t77\t0.0\t1\t75 2\n",
      "99\t48\t0.0\t1\t48\n",
      "100\t56\t0.0\t1\t55 1\n",
      "101\t52\t0.0\t1\t52\n",
      "102\t49\t0.0\t1\t49\n",
      "103\t60\t0.0\t1\t59 1\n",
      "104\t76\t0.0\t1\t76\n",
      "105\t62\t0.0\t1\t60 2\n",
      "106\t74\t0.0\t1\t74\n",
      "107\t58\t0.0\t1\t58\n",
      "108\t68\t0.0\t1\t67 1\n",
      "109\t56\t0.0\t1\t56\n",
      "110\t50\t0.0\t1\t49 1\n",
      "111\t43\t0.0\t1\t42 1\n",
      "112\t34\t0.0\t1\t34\n",
      "113\t35\t0.0\t1\t33 2\n",
      "114\t36\t0.0\t1\t33 3\n",
      "115\t15\t0.0\t1\t15\n",
      "116\t15\t0.0\t1\t15\n",
      "117\t7\t0.0\t1\t6 1\n",
      "118\t4\t0.0\t1\t4\n",
      "119\t2\t0.0\t1\t2\n",
      "121\t2\t0.0\t1\t2\n",
      "122\t6\t0.0\t1\t5 1\n",
      "129\t1\t0.0\t1\t0 1\n",
      "137\t2\t0.0\t1\t2\n",
      "139\t1\t0.0\t1\t1\n",
      "149\t1\t0.0\t1\t0 1\n",
      "150\t4\t0.0\t1\t4\n",
      "\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /scratch/hansetj1/script_testing/fastq_processing/raw_reads/sub_read1.fq\n",
      "=============================================\n",
      "10000 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Writing report to '/gpfs23/scratch/hansetj1/script_testing/fastq_processing/trimmed_reads/sub_read2.fq_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /scratch/hansetj1/script_testing/fastq_processing/raw_reads/sub_read2.fq\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.6\n",
      "Cutadapt version: 1.18\n",
      "Number of cores used for trimming: 1\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'CTGTCTCTTATA' (Nextera Transposase sequence; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Running FastQC on the data once trimming has completed\n",
      "Running FastQC with the following extra arguments: '--outdir /scratch/hansetj1/script_testing/fastq_processing/qc'\n",
      "\n",
      "Cutadapt seems to be reasonably up-to-date. Setting -j -j 1\n",
      "Writing final adapter and quality trimmed output to sub_read2_trimmed.fq\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'CTGTCTCTTATA' from file /scratch/hansetj1/script_testing/fastq_processing/raw_reads/sub_read2.fq <<< \n",
      "This is cutadapt 1.18 with Python 3.6.13\n",
      "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a CTGTCTCTTATA /scratch/hansetj1/script_testing/fastq_processing/raw_reads/sub_read2.fq\n",
      "Processing reads on 1 core in single-end mode ...\n",
      "Finished in 0.17 s (17 us/read; 3.63 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:                  10,000\n",
      "Reads with adapters:                     7,330 (73.3%)\n",
      "Reads written (passing filters):        10,000 (100.0%)\n",
      "\n",
      "Total basepairs processed:     1,496,688 bp\n",
      "Quality-trimmed:                   6,421 bp (0.4%)\n",
      "Total written (filtered):      1,156,722 bp (77.3%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: CTGTCTCTTATA; Type: regular 3'; Length: 12; Trimmed: 7330 times.\n",
      "\n",
      "No. of allowed errors:\n",
      "0-9 bp: 0; 10-12 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 13.5%\n",
      "  C: 40.1%\n",
      "  G: 26.4%\n",
      "  T: 19.9%\n",
      "  none/other: 0.1%\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t1652\t2500.0\t0\t1652\n",
      "2\t449\t625.0\t0\t449\n",
      "3\t152\t156.2\t0\t152\n",
      "4\t55\t39.1\t0\t55\n",
      "5\t41\t9.8\t0\t41\n",
      "6\t32\t2.4\t0\t32\n",
      "7\t26\t0.6\t0\t26\n",
      "8\t31\t0.2\t0\t31\n",
      "9\t43\t0.0\t0\t43\n",
      "11\t50\t0.0\t1\t48 2\n",
      "13\t57\t0.0\t1\t55 2\n",
      "14\t1\t0.0\t1\t1\n",
      "15\t79\t0.0\t1\t79\n",
      "16\t2\t0.0\t1\t2\n",
      "17\t26\t0.0\t1\t25 1\n",
      "18\t30\t0.0\t1\t27 3\n",
      "19\t29\t0.0\t1\t29\n",
      "20\t62\t0.0\t1\t60 2\n",
      "21\t1\t0.0\t1\t0 1\n",
      "22\t30\t0.0\t1\t28 2\n",
      "23\t27\t0.0\t1\t26 1\n",
      "24\t31\t0.0\t1\t31\n",
      "25\t43\t0.0\t1\t42 1\n",
      "26\t28\t0.0\t1\t27 1\n",
      "27\t37\t0.0\t1\t37\n",
      "28\t39\t0.0\t1\t39\n",
      "29\t62\t0.0\t1\t61 1\n",
      "30\t1\t0.0\t1\t0 1\n",
      "31\t31\t0.0\t1\t31\n",
      "32\t75\t0.0\t1\t73 2\n",
      "33\t2\t0.0\t1\t1 1\n",
      "34\t39\t0.0\t1\t38 1\n",
      "35\t38\t0.0\t1\t38\n",
      "36\t30\t0.0\t1\t30\n",
      "37\t36\t0.0\t1\t33 3\n",
      "38\t38\t0.0\t1\t36 2\n",
      "39\t62\t0.0\t1\t60 2\n",
      "40\t5\t0.0\t1\t4 1\n",
      "41\t48\t0.0\t1\t48\n",
      "42\t42\t0.0\t1\t39 3\n",
      "43\t39\t0.0\t1\t39\n",
      "44\t67\t0.0\t1\t65 2\n",
      "45\t205\t0.0\t1\t200 5\n",
      "46\t3\t0.0\t1\t3\n",
      "47\t16\t0.0\t1\t16\n",
      "48\t2\t0.0\t1\t2\n",
      "49\t18\t0.0\t1\t17 1\n",
      "50\t4\t0.0\t1\t4\n",
      "51\t39\t0.0\t1\t39\n",
      "52\t23\t0.0\t1\t23\n",
      "53\t58\t0.0\t1\t58\n",
      "54\t61\t0.0\t1\t61\n",
      "55\t28\t0.0\t1\t28\n",
      "56\t48\t0.0\t1\t46 2\n",
      "57\t140\t0.0\t1\t139 1\n",
      "58\t35\t0.0\t1\t34 1\n",
      "59\t22\t0.0\t1\t20 2\n",
      "60\t99\t0.0\t1\t98 1\n",
      "61\t29\t0.0\t1\t29\n",
      "62\t35\t0.0\t1\t35\n",
      "63\t167\t0.0\t1\t166 1\n",
      "64\t103\t0.0\t1\t102 1\n",
      "65\t11\t0.0\t1\t11\n",
      "66\t52\t0.0\t1\t50 2\n",
      "67\t71\t0.0\t1\t70 1\n",
      "68\t12\t0.0\t1\t11 1\n",
      "69\t23\t0.0\t1\t22 1\n",
      "70\t74\t0.0\t1\t74\n",
      "71\t32\t0.0\t1\t32\n",
      "72\t24\t0.0\t1\t23 1\n",
      "73\t36\t0.0\t1\t33 3\n",
      "74\t33\t0.0\t1\t33\n",
      "75\t44\t0.0\t1\t44\n",
      "76\t46\t0.0\t1\t44 2\n",
      "77\t40\t0.0\t1\t39 1\n",
      "78\t41\t0.0\t1\t41\n",
      "79\t41\t0.0\t1\t41\n",
      "80\t39\t0.0\t1\t39\n",
      "81\t47\t0.0\t1\t47\n",
      "82\t50\t0.0\t1\t50\n",
      "83\t60\t0.0\t1\t58 2\n",
      "84\t59\t0.0\t1\t58 1\n",
      "85\t56\t0.0\t1\t55 1\n",
      "86\t59\t0.0\t1\t58 1\n",
      "87\t50\t0.0\t1\t50\n",
      "88\t58\t0.0\t1\t58\n",
      "89\t53\t0.0\t1\t53\n",
      "90\t55\t0.0\t1\t55\n",
      "91\t57\t0.0\t1\t57\n",
      "92\t54\t0.0\t1\t54\n",
      "93\t43\t0.0\t1\t43\n",
      "94\t50\t0.0\t1\t49 1\n",
      "95\t58\t0.0\t1\t57 1\n",
      "96\t69\t0.0\t1\t69\n",
      "97\t59\t0.0\t1\t59\n",
      "98\t75\t0.0\t1\t75\n",
      "99\t55\t0.0\t1\t55\n",
      "100\t57\t0.0\t1\t54 3\n",
      "101\t54\t0.0\t1\t53 1\n",
      "102\t43\t0.0\t1\t43\n",
      "103\t55\t0.0\t1\t54 1\n",
      "104\t70\t0.0\t1\t69 1\n",
      "105\t60\t0.0\t1\t60\n",
      "106\t74\t0.0\t1\t74\n",
      "107\t53\t0.0\t1\t53\n",
      "108\t67\t0.0\t1\t66 1\n",
      "109\t53\t0.0\t1\t53\n",
      "110\t48\t0.0\t1\t48\n",
      "111\t41\t0.0\t1\t38 3\n",
      "112\t35\t0.0\t1\t35\n",
      "113\t36\t0.0\t1\t35 1\n",
      "114\t34\t0.0\t1\t34\n",
      "115\t14\t0.0\t1\t14\n",
      "116\t14\t0.0\t1\t14\n",
      "117\t5\t0.0\t1\t5\n",
      "118\t4\t0.0\t1\t4\n",
      "119\t2\t0.0\t1\t2\n",
      "121\t2\t0.0\t1\t1 1\n",
      "122\t7\t0.0\t1\t5 2\n",
      "125\t1\t0.0\t1\t0 1\n",
      "126\t1\t0.0\t1\t0 1\n",
      "132\t1\t0.0\t1\t0 1\n",
      "135\t1\t0.0\t1\t0 1\n",
      "137\t2\t0.0\t1\t2\n",
      "139\t1\t0.0\t1\t1\n",
      "147\t1\t0.0\t1\t0 1\n",
      "149\t1\t0.0\t1\t0 1\n",
      "150\t4\t0.0\t1\t4\n",
      "\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /scratch/hansetj1/script_testing/fastq_processing/raw_reads/sub_read2.fq\n",
      "=============================================\n",
      "10000 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files sub_read1_trimmed.fq and sub_read2_trimmed.fq\n",
      "Now renaming the output files\n",
      "\n",
      "ORIGINAL FILE 1: >>sub_read1_trimmed.fq<<\tRENAMING TO:>>test_set_R1_trimmed.fq<<\n",
      "ORIGINAL FILE 2: >>sub_read2_trimmed.fq<<\tRENAMING TO:>>test_set_R2_trimmed.fq<<\n",
      "file_1: test_set_R1_trimmed.fq, file_2: test_set_R2_trimmed.fq\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: test_set_R1_trimmed.fq and test_set_R2_trimmed.fq <<<<<\n",
      "Renaming the output files (AGAIN).\n",
      "Writing validated paired-end Read 1 reads to test_set_val_1.fq\n",
      "Writing validated paired-end Read 2 reads to test_set_val_2.fq\n",
      "\n",
      "Total number of sequences analysed: 10000\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 12 (0.12%)\n",
      "\n",
      "\n",
      "  >>> Now running FastQC on the validated data test_set_val_1.fq<<<\n",
      "\n",
      "Started analysis of test_set_val_1.fq\n",
      "Approx 10% complete for test_set_val_1.fq\n",
      "Approx 20% complete for test_set_val_1.fq\n",
      "Approx 30% complete for test_set_val_1.fq\n",
      "Approx 40% complete for test_set_val_1.fq\n",
      "Approx 50% complete for test_set_val_1.fq\n",
      "Approx 60% complete for test_set_val_1.fq\n",
      "Approx 70% complete for test_set_val_1.fq\n",
      "Approx 80% complete for test_set_val_1.fq\n",
      "Approx 90% complete for test_set_val_1.fq\n",
      "Analysis complete for test_set_val_1.fq\n",
      "\n",
      "  >>> Now running FastQC on the validated data test_set_val_2.fq<<<\n",
      "\n",
      "Started analysis of test_set_val_2.fq\n",
      "Approx 10% complete for test_set_val_2.fq\n",
      "Approx 20% complete for test_set_val_2.fq\n",
      "Approx 30% complete for test_set_val_2.fq\n",
      "Approx 40% complete for test_set_val_2.fq\n",
      "Approx 50% complete for test_set_val_2.fq\n",
      "Approx 60% complete for test_set_val_2.fq\n",
      "Approx 70% complete for test_set_val_2.fq\n",
      "Approx 80% complete for test_set_val_2.fq\n",
      "Approx 90% complete for test_set_val_2.fq\n",
      "Analysis complete for test_set_val_2.fq\n",
      "Deleting both intermediate output files test_set_R1_trimmed.fq and test_set_R2_trimmed.fq\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "\n",
      "***Mapping trimmed reads\n",
      "9988 reads; of these:\n",
      "  9988 (100.00%) were paired; of these:\n",
      "    721 (7.22%) aligned concordantly 0 times\n",
      "    4821 (48.27%) aligned concordantly exactly 1 time\n",
      "    4446 (44.51%) aligned concordantly >1 times\n",
      "92.78% overall alignment rate\n",
      "\n",
      "***Filtering mapped reads\n",
      "[bam_sort_core] merging from 0 files and 8 in-memory blocks...\n",
      "INFO\t2021-10-18 17:09:53\tMarkDuplicates\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    MarkDuplicates -I /scratch/hansetj1/script_testing/fastq_processing/bams/test_set.filtered.pos-sorted.bam -O /scratch/hansetj1/script_testing/fastq_processing/bams/test_set.unique.unsorted.bam -M /scratch/hansetj1/script_testing/fastq_processing/qc/test_set_marked_dup_metrics-all.txt -REMOVE_DUPLICATES TRUE\n",
      "**********\n",
      "\n",
      "\n",
      "17:09:53.725 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/gpfs52/home/hansetj1/.conda/envs/jupyter/share/picard-2.26.3-0/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Mon Oct 18 17:09:53 CDT 2021] MarkDuplicates INPUT=[/scratch/hansetj1/script_testing/fastq_processing/bams/test_set.filtered.pos-sorted.bam] OUTPUT=/scratch/hansetj1/script_testing/fastq_processing/bams/test_set.unique.unsorted.bam METRICS_FILE=/scratch/hansetj1/script_testing/fastq_processing/qc/test_set_marked_dup_metrics-all.txt REMOVE_DUPLICATES=true    MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP=50000 MAX_FILE_HANDLES_FOR_READ_ENDS_MAP=8000 SORTING_COLLECTION_SIZE_RATIO=0.25 TAG_DUPLICATE_SET_MEMBERS=false REMOVE_SEQUENCING_DUPLICATES=false TAGGING_POLICY=DontTag CLEAR_DT=true DUPLEX_UMI=false ADD_PG_TAG_TO_READS=true ASSUME_SORTED=false DUPLICATE_SCORING_STRATEGY=SUM_OF_BASE_QUALITIES PROGRAM_RECORD_ID=MarkDuplicates PROGRAM_GROUP_NAME=MarkDuplicates READ_NAME_REGEX=<optimized capture of last three ':' separated fields as numeric values> OPTICAL_DUPLICATE_PIXEL_DISTANCE=100 MAX_OPTICAL_DUPLICATE_SET_SIZE=300000 VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false GA4GH_CLIENT_SECRETS=client_secrets.json USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
      "[Mon Oct 18 17:09:53 CDT 2021] Executing as hansetj1@cn1296.vampire on Linux 3.10.0-1160.36.2.el7.x86_64 amd64; OpenJDK 64-Bit Server VM 1.8.0_302-b08; Deflater: Intel; Inflater: Intel; Provider GCS is not available; Picard version: 2.26.3\n",
      "INFO\t2021-10-18 17:09:53\tMarkDuplicates\tStart of doWork freeMemory: 1117100784; totalMemory: 1133510656; maxMemory: 16799760384\n",
      "INFO\t2021-10-18 17:09:53\tMarkDuplicates\tReading input file and constructing read end information.\n",
      "INFO\t2021-10-18 17:09:53\tMarkDuplicates\tWill retain up to 60868697 data points before spilling to disk.\n",
      "INFO\t2021-10-18 17:09:54\tMarkDuplicates\tRead 0 records. 0 pairs never matched.\n",
      "INFO\t2021-10-18 17:09:55\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 636180192; totalMemory: 1133510656; maxMemory: 16799760384\n",
      "INFO\t2021-10-18 17:09:55\tMarkDuplicates\tWill retain up to 524992512 duplicate indices before spilling to disk.\n",
      "INFO\t2021-10-18 17:09:57\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
      "INFO\t2021-10-18 17:09:57\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
      "INFO\t2021-10-18 17:09:57\tMarkDuplicates\tSorting list of duplicate records.\n",
      "INFO\t2021-10-18 17:09:57\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 1123811472; totalMemory: 5333581824; maxMemory: 16799760384\n",
      "INFO\t2021-10-18 17:09:57\tMarkDuplicates\tMarking 0 records as duplicates.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO\t2021-10-18 17:09:57\tMarkDuplicates\tFound 0 optical duplicate clusters.\n",
      "INFO\t2021-10-18 17:09:57\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
      "INFO\t2021-10-18 17:09:57\tMarkDuplicates\tWriting complete. Closing input iterator.\n",
      "INFO\t2021-10-18 17:09:57\tMarkDuplicates\tDuplicate Index cleanup.\n",
      "INFO\t2021-10-18 17:09:57\tMarkDuplicates\tGetting Memory Stats.\n",
      "INFO\t2021-10-18 17:09:57\tMarkDuplicates\tBefore output close freeMemory: 5323121096; totalMemory: 5333581824; maxMemory: 16799760384\n",
      "INFO\t2021-10-18 17:09:57\tMarkDuplicates\tClosed outputs. Getting more Memory Stats.\n",
      "INFO\t2021-10-18 17:09:57\tMarkDuplicates\tAfter output close freeMemory: 5323196224; totalMemory: 5333581824; maxMemory: 16799760384\n",
      "[Mon Oct 18 17:09:57 CDT 2021] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 0.07 minutes.\n",
      "Runtime.totalMemory()=5333581824\n",
      "\n",
      "***Counting mapped reads\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "dir='/scratch/hansetj1/script_testing/fastq_processing'\n",
    "\n",
    "python3 ./fastq_processing/fastq_processing.py -a ${dir}/raw_reads/sub_read1.fq -b ${dir}/raw_reads/sub_read2.fq \\\n",
    "    -n test_set -e /home/hansetj1/hg38_encode_blacklist_ENCFF356LFX.bed -t ${dir}/trimmed_reads \\\n",
    "    -o ${dir}/bams -r ${dir}/qc -g /data/hodges_lab/hg38_genome/bt2/hg38"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "431b8a78",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 5.8M\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 2.9M Oct 18 17:09 test_set_val_1.fq\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 2.9M Oct 18 17:09 test_set_val_2.fq\n",
      "total 11M\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 7.2M Oct 18 17:09 test_set.sam\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 817K Oct 18 17:09 test_set.bam\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 1.5M Oct 18 17:09 test_set.bam.bai\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 287K Oct 18 17:09 test_set.chrM.bam\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 7.0K Oct 18 17:09 test_set.no_chrM.unsorted.bam\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 7.1K Oct 18 17:09 test_set.no_chrM.bam\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 3.6K Oct 18 17:09 test_set.no_chrM.bam.bai\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 7.1K Oct 18 17:09 test_set.blacklisted.bam\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 7.1K Oct 18 17:09 test_set.filtered.unsorted.bam\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 7.1K Oct 18 17:09 test_set.filtered.pos-sorted.bam\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 3.6K Oct 18 17:09 test_set.filtered.pos-sorted.bam.bai\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 7.2K Oct 18 17:09 test_set.filtered.n-sorted.bam\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 7.5K Oct 18 17:09 test_set.unique.unsorted.bam\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 7.2K Oct 18 17:09 test_set.unique.pos-sorted.bam\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 3.6K Oct 18 17:09 test_set.unique.pos-sorted.bam.bai\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 7.2K Oct 18 17:09 test_set.unique.n-sorted.bam\n",
      "total 1.4M\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 3.9K Oct 18 17:09 sub_read1.fq_trimming_report.txt\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 4.1K Oct 18 17:09 sub_read2.fq_trimming_report.txt\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 372K Oct 18 17:09 test_set_val_1_fastqc.zip\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 262K Oct 18 17:09 test_set_val_1_fastqc.html\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 376K Oct 18 17:09 test_set_val_2_fastqc.zip\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 261K Oct 18 17:09 test_set_val_2_fastqc.html\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 1.3K Oct 18 17:09 test_set_marked_dup_metrics-all.txt\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab  488 Oct 18 17:09 test_set.sam_flagstat-results.txt\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab  491 Oct 18 17:09 test_set.bam_flagstat-results.txt\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab  443 Oct 18 17:09 test_set.no_chrM.bam_flagstat-results.txt\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab  443 Oct 18 17:09 test_set.filtered.pos-sorted.bam_flagstat-results.txt\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab  443 Oct 18 17:09 test_set.unique.pos-sorted.bam_flagstat-results.txt\n"
     ]
    }
   ],
   "source": [
    "#check output files\n",
    "\n",
    "dir='/scratch/hansetj1/script_testing/fastq_processing'\n",
    "\n",
    "ls -lhtr ${dir}/trimmed_reads\n",
    "ls -lhtr ${dir}/bams\n",
    "ls -lhtr ${dir}/qc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "94617438",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A00758:23:HH7TCDSXX:2:2636:4878:13479\t99\tchrM\t1\t40\t4M1I145M\t=\t193\t340\tGGATCACAGGTCTATCACCCTATTAACCACTCACGGGAGCTCTCCATGCATTTGGTATTTTCGTCTGGGGGGTATGCACGCGATAGCATTGCGAGACGCTGGAGCCGGAGCACCCTATGTCGCAGTATCTGTCTTTGATTCCTGCCTCAT\tFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\tAS:i:-23\tXN:i:0\tXM:i:3\tXO:i:1\tXG:i:1\tNM:i:4\tMD:Z:1A0T0C145\tYS:i:-13\tYT:Z:CP\n",
      "A00758:23:HH7TCDSXX:2:1124:13530:23015\t99\tchrM\t1\t38\t119M\t=\t1\t119\tGATCACAGGTCTATCACCCTATTAACCACTCACGGGAGCTCTCCATGCATTTGGTATTTTCGTCTGGGGGGTATGCACGCGATAGCATTGCGAGACGCTGGAGCCGGAGCACCCTATGT\tFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF::FFF:,FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF,FFF\tAS:i:0\tXS:i:-60\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:119\tYS:i:0\tYT:Z:CP\n",
      "A00758:23:HH7TCDSXX:2:2565:11614:30185\t99\tchrM\t1\t40\t4M5I140M\t=\t55\t203\tCGATGGATCACAGGTCTATCACCCTATTAACCACTCACGGGAGCTCTCCATGCATTTGGTATTTTCGTCTGGGGGGTATGCACGCGATAGCATTGCGAGACGCTGGAGCCGGAGCACCCTATGTCGCAGTATCTGTCTTTGATTCCTGC\tFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFF:FFFFFFFFFFFFFFFFF\tAS:i:-40\tXN:i:0\tXM:i:4\tXO:i:1\tXG:i:5\tNM:i:9\tMD:Z:0G0A0T0C140\tYS:i:0\tYT:Z:CP\n",
      "A00758:23:HH7TCDSXX:2:2326:29387:15718\t163\tchrM\t1\t36\t86M\t=\t1\t86\tGATCACAGGTCTATCACCCTATTAACCACTCACGGGAGCTCTCCATGCATTTGGTATTTTCGTCTGGGGGGTATGCACGCGATAGC\tFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFF\tAS:i:0\tXS:i:-35\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:86\tYS:i:0\tYT:Z:CP\n",
      "A00758:23:HH7TCDSXX:2:1124:13530:23015\t147\tchrM\t1\t38\t119M\t=\t1\t-119\tGATCACAGGTCTATCACCCTATTAACCACTCACGGGAGCTCTCCATGCATTTGGTATTTTCGTCTGGGGGGTATGCACGCGATAGCATTGCGAGACGCTGGAGCCGGAGCACCCTATGT\tF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\tAS:i:0\tXS:i:-60\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:119\tYS:i:0\tYT:Z:CP\n",
      "A00758:23:HH7TCDSXX:2:2326:29387:15718\t83\tchrM\t1\t36\t86M\t=\t1\t-86\tGATCACAGGTCTATCACCCTATTAACCACTCACGGGAGCTCTCCATGCATTTGGTATTTTCGTCTGGGGGGTATGCACGCGATAGC\tF:FFFFFFFFFFFFFFFFFFFFFF:FF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\tAS:i:0\tXS:i:-35\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:86\tYS:i:0\tYT:Z:CP\n",
      "A00758:23:HH7TCDSXX:2:1277:11342:12806\t99\tchrM\t8\t38\t110M\t=\t8\t110\tGGTCTATCACCCTATTAACCACTCACGGGAGCTCTCCATGCATTTGGTATTTTCGTCTGGGGGGTATGCACGCGATAGCATTGCGAGACGCTGGAGCCGGAGCACCCTAT\tF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF,FFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\tAS:i:0\tXS:i:-55\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:110\tYS:i:0\tYT:Z:CP\n",
      "A00758:23:HH7TCDSXX:2:1277:11342:12806\t147\tchrM\t8\t38\t110M\t=\t8\t-110\tGGTCTATCACCCTATTAACCACTCACGGGAGCTCTCCATGCATTTGGTATTTTCGTCTGGGGGGTATGCACGCGATAGCATTGCGAGACGCTGGAGCCGGAGCACCCTAT\tFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:,\tAS:i:0\tXS:i:-55\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:110\tYS:i:0\tYT:Z:CP\n",
      "A00758:23:HH7TCDSXX:2:2275:31982:8343\t163\tchrM\t9\t42\t120M\t=\t9\t120\tGTCTATCACCCTATTAACCACTCACGGGAGCTCTCCATGCATTTGGTATTTTCGTCTGGGGGGTATGCACGCGATAGCATTGCGAGACGCTGGAGCCGGAGCACCCTATGTCGCAGTATC\tFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\tAS:i:0\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:120\tYS:i:0\tYT:Z:CP\n",
      "A00758:23:HH7TCDSXX:2:1267:17083:30890\t163\tchrM\t9\t39\t102M\t=\t9\t102\tGTCTATCACCCTATTAACCACTCACGGGAGCTCTCCATGCATTTGGTATTTTCGTCTGGGGGGTATGCACGCGATAGCATTGCGAGACGCTGGAGCCGGAGC\tFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\tAS:i:0\tXS:i:-55\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:102\tYS:i:0\tYT:Z:CP\n",
      "samtools view: writing to standard output failed: Broken pipe\n",
      "samtools view: error closing standard output: -1\n"
     ]
    }
   ],
   "source": [
    "#It looks like there was an error in the chrM step. Check the bam files in both the chrM and no_chrM\n",
    "\n",
    "samtools view ${dir}/bams/test_set.chrM.bam | head\n",
    "samtools view ${dir}/bams/test_set.no_chrM.bam | head"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c07b865",
   "metadata": {},
   "source": [
    "### Problem with no output to no_ChrM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a8af19f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#It looks like no reads were written to th no_ChrM file. Try to fix that. \n",
    "\n",
    "bam_dir='/scratch/hansetj1/script_testing/fastq_processing/bams'\n",
    "\n",
    "samtools view -@ 8 -h ${bam_dir}/test_set.bam | awk '{if($3 != \"chrM\"){print $0}}' - | \\\n",
    "    samtools view -S -b - > ${bam_dir}/test_set.no_chrM.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "83279098",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A00758:23:HH7TCDSXX:2:2271:23348:18850\t99\tchr1\t629912\t30\t80M\t=\t629912\t80\tTCAATACTCATCATTAATAATCATAATGGCTATAGCAATAAAACTAGGAATAGCCCCCTTTCACTTCTGAGTCCCAGAGG\tFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\tAS:i:0\tXS:i:-5\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:80\tYS:i:0\tYT:Z:CP\n",
      "A00758:23:HH7TCDSXX:2:2271:23348:18850\t147\tchr1\t629912\t30\t80M\t=\t629912\t-80\tTCAATACTCATCATTAATAATCATAATGGCTATAGCAATAAAACTAGGAATAGCCCCCTTTCACTTCTGAGTCCCAGAGG\tFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\tAS:i:0\tXS:i:-5\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:80\tYS:i:0\tYT:Z:CP\n",
      "A00758:23:HH7TCDSXX:2:2211:8034:35415\t163\tchr1\t629922\t30\t70M\t=\t629922\t70\tTCATTAATAATCATAATGGCTATAGCAATAAAACTAGGAATAGCCCCCTTTCACTTCTGAGTCCCAGAGG\tFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFF\tAS:i:0\tXS:i:-5\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:70\tYS:i:0\tYT:Z:CP\n",
      "A00758:23:HH7TCDSXX:2:2211:8034:35415\t83\tchr1\t629922\t30\t70M\t=\t629922\t-70\tTCATTAATAATCATAATGGCTATAGCAATAAAACTAGGAATAGCCCCCTTTCACTTCTGAGTCCCAGAGG\tFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFF\tAS:i:0\tXS:i:-5\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:70\tYS:i:0\tYT:Z:CP\n",
      "A00758:23:HH7TCDSXX:2:1511:11885:11334\t163\tchr1\t629939\t30\t79M\t=\t629939\t79\tGGCTATAGCAATAAAACTAGGAATAGCCCCCTTTCACTTCTGAGTCCCAGAGGTTACCCAAGGCACCCCTCTGACATCC\tFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF,FFF\tAS:i:0\tXS:i:-5\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:79\tYS:i:0\tYT:Z:CP\n",
      "A00758:23:HH7TCDSXX:2:1511:11885:11334\t83\tchr1\t629939\t30\t79M\t=\t629939\t-79\tGGCTATAGCAATAAAACTAGGAATAGCCCCCTTTCACTTCTGAGTCCCAGAGGTTACCCAAGGCACCCCTCTGACATCC\tFFFFFFFFFF:FFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\tAS:i:0\tXS:i:-5\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:79\tYS:i:0\tYT:Z:CP\n",
      "A00758:23:HH7TCDSXX:2:2444:24026:22216\t163\tchr1\t633991\t30\t73M\t=\t633991\t73\tCTATAAACCTAGCCATGGCCATCCCCTTATGAGCGGGCGCAGTGATTATAGGCTTTCGCTCTAAGATTAAAAA\tFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFF\tAS:i:0\tXS:i:-5\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:73\tYS:i:0\tYT:Z:CP\n",
      "A00758:23:HH7TCDSXX:2:2568:7609:28072\t99\tchr1\t633991\t30\t52M\t=\t633991\t52\tCTATAAACCTAGCCATGGCCATCCCCTTATGAGCGGGCGCAGTGATTATAGG\tFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\tAS:i:0\tXS:i:-5\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:52\tYS:i:0\tYT:Z:CP\n",
      "A00758:23:HH7TCDSXX:2:2422:27859:33019\t163\tchr1\t633991\t30\t45M\t=\t633991\t45\tCTATAAACCTAGCCATGGCCATCCCCTTATGAGCGGGCGCAGTGA\tFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\tAS:i:0\tXS:i:-5\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:45\tYS:i:0\tYT:Z:CP\n",
      "A00758:23:HH7TCDSXX:2:2444:24026:22216\t83\tchr1\t633991\t30\t73M\t=\t633991\t-73\tCTATAAACCTAGCCATGGCCATCCCCTTATGAGCGGGCGCAGTGATTATAGGCTTTCGCTCTAAGATTAAAAA\tFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\tAS:i:0\tXS:i:-5\tXN:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tNM:i:0\tMD:Z:73\tYS:i:0\tYT:Z:CP\n",
      "grep: write error\n",
      "samtools view: writing to standard output failed: Broken pipe\n",
      "samtools view: error closing standard output: -1\n"
     ]
    }
   ],
   "source": [
    "samtools view ${bam_dir}/test_set.no_chrM.bam | grep chr1 | head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c45ede96",
   "metadata": {},
   "outputs": [
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "samtools view ${bam_dir}/test_set.no_chrM.bam | grep chrM "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c22350f",
   "metadata": {},
   "source": [
    "### Solution is to use awk\n",
    "I fixed the issue by using awk to filter rather than samtools -view options. It works better and is simpler. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01713886",
   "metadata": {},
   "source": [
    "### Redo of run-through with update"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "4dc92abb",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir='/scratch/hansetj1/script_testing/fastq_processing'\n",
    "\n",
    "rm ${dir}/trimmed_reads/*\n",
    "rm ${dir}/bams/*\n",
    "rm ${dir}/qc/*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "20f5a9be",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PARAMETERS:\n",
      "Fastq R1: /scratch/hansetj1/script_testing/fastq_processing/raw_reads/sub_read1.fq\n",
      "Fastq R2: /scratch/hansetj1/script_testing/fastq_processing/raw_reads/sub_read2.fq\n",
      "Sample name: test_set_2\n",
      "Blacklist Bedfile /home/hansetj1/hg38_encode_blacklist_ENCFF356LFX.bed\n",
      "Trimmed reads directory: /scratch/hansetj1/script_testing/fastq_processing/trimmed_reads\n",
      "Mapped reads directory: /scratch/hansetj1/script_testing/fastq_processing/bams\n",
      "QC results directory: /scratch/hansetj1/script_testing/fastq_processing/qc\n",
      "Genome: /data/hodges_lab/hg38_genome/bt2/hg38\n",
      "Detected and using 8 cores for processing\n",
      "\n",
      "***Trimming raw reads\n",
      "Multicore support not enabled. Proceeding with single-core trimming.\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 1.18\n",
      "single-core operation.\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output will be written into the directory: /gpfs23/scratch/hansetj1/script_testing/fastq_processing/trimmed_reads/\n",
      "Using user-specified basename (>>test_set_2<<) instead of deriving the filename from the input file(s)\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> /scratch/hansetj1/script_testing/fastq_processing/raw_reads/sub_read1.fq <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "Nextera\t4704\tCTGTCTCTTATA\t10000\t47.04\n",
      "smallRNA\t0\tTGGAATTCTCGG\t10000\t0.00\n",
      "Illumina\t0\tAGATCGGAAGAGC\t10000\t0.00\n",
      "Using Nextera adapter for trimming (count: 4704). Second best hit was smallRNA (count: 0)\n",
      "\n",
      "Writing report to '/gpfs23/scratch/hansetj1/script_testing/fastq_processing/trimmed_reads/sub_read1.fq_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /scratch/hansetj1/script_testing/fastq_processing/raw_reads/sub_read1.fq\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.6\n",
      "Cutadapt version: 1.18\n",
      "Number of cores used for trimming: 1\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'CTGTCTCTTATA' (Nextera Transposase sequence; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Running FastQC on the data once trimming has completed\n",
      "Running FastQC with the following extra arguments: '--outdir /scratch/hansetj1/script_testing/fastq_processing/qc'\n",
      "\n",
      "Cutadapt seems to be reasonably up-to-date. Setting -j 1\n",
      "Writing final adapter and quality trimmed output to sub_read1_trimmed.fq\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'CTGTCTCTTATA' from file /scratch/hansetj1/script_testing/fastq_processing/raw_reads/sub_read1.fq <<< \n",
      "This is cutadapt 1.18 with Python 3.6.13\n",
      "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a CTGTCTCTTATA /scratch/hansetj1/script_testing/fastq_processing/raw_reads/sub_read1.fq\n",
      "Processing reads on 1 core in single-end mode ...\n",
      "Finished in 0.17 s (17 us/read; 3.43 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:                  10,000\n",
      "Reads with adapters:                     7,375 (73.8%)\n",
      "Reads written (passing filters):        10,000 (100.0%)\n",
      "\n",
      "Total basepairs processed:     1,496,800 bp\n",
      "Quality-trimmed:                   8,932 bp (0.6%)\n",
      "Total written (filtered):      1,157,015 bp (77.3%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: CTGTCTCTTATA; Type: regular 3'; Length: 12; Trimmed: 7375 times.\n",
      "\n",
      "No. of allowed errors:\n",
      "0-9 bp: 0; 10-12 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 13.4%\n",
      "  C: 40.3%\n",
      "  G: 25.5%\n",
      "  T: 20.7%\n",
      "  none/other: 0.1%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t1710\t2500.0\t0\t1710\n",
      "2\t426\t625.0\t0\t426\n",
      "3\t162\t156.2\t0\t162\n",
      "4\t56\t39.1\t0\t56\n",
      "5\t34\t9.8\t0\t34\n",
      "6\t34\t2.4\t0\t34\n",
      "7\t24\t0.6\t0\t24\n",
      "8\t34\t0.2\t0\t34\n",
      "9\t46\t0.0\t0\t46\n",
      "11\t48\t0.0\t1\t47 1\n",
      "12\t1\t0.0\t1\t1\n",
      "13\t57\t0.0\t1\t57\n",
      "14\t1\t0.0\t1\t0 1\n",
      "15\t79\t0.0\t1\t77 2\n",
      "16\t3\t0.0\t1\t2 1\n",
      "17\t27\t0.0\t1\t27\n",
      "18\t28\t0.0\t1\t28\n",
      "19\t27\t0.0\t1\t25 2\n",
      "20\t39\t0.0\t1\t39\n",
      "21\t30\t0.0\t1\t29 1\n",
      "22\t85\t0.0\t1\t85\n",
      "24\t1\t0.0\t1\t0 1\n",
      "25\t43\t0.0\t1\t42 1\n",
      "26\t27\t0.0\t1\t26 1\n",
      "27\t80\t0.0\t1\t79 1\n",
      "29\t37\t0.0\t1\t35 2\n",
      "30\t129\t0.0\t1\t128 1\n",
      "31\t1\t0.0\t1\t0 1\n",
      "33\t1\t0.0\t1\t0 1\n",
      "34\t39\t0.0\t1\t39\n",
      "35\t67\t0.0\t1\t66 1\n",
      "36\t60\t0.0\t1\t60\n",
      "37\t2\t0.0\t1\t0 2\n",
      "38\t1\t0.0\t1\t1\n",
      "39\t27\t0.0\t1\t27\n",
      "40\t41\t0.0\t1\t38 3\n",
      "41\t89\t0.0\t1\t88 1\n",
      "42\t45\t0.0\t1\t44 1\n",
      "43\t6\t0.0\t1\t6\n",
      "44\t51\t0.0\t1\t49 2\n",
      "45\t110\t0.0\t1\t106 4\n",
      "46\t14\t0.0\t1\t14\n",
      "47\t45\t0.0\t1\t44 1\n",
      "48\t292\t0.0\t1\t289 3\n",
      "49\t55\t0.0\t1\t55\n",
      "50\t25\t0.0\t1\t23 2\n",
      "51\t168\t0.0\t1\t165 3\n",
      "52\t41\t0.0\t1\t41\n",
      "53\t5\t0.0\t1\t4 1\n",
      "54\t72\t0.0\t1\t72\n",
      "55\t64\t0.0\t1\t63 1\n",
      "56\t28\t0.0\t1\t28\n",
      "57\t21\t0.0\t1\t20 1\n",
      "58\t7\t0.0\t1\t7\n",
      "59\t46\t0.0\t1\t46\n",
      "60\t6\t0.0\t1\t6\n",
      "61\t13\t0.0\t1\t13\n",
      "62\t37\t0.0\t1\t36 1\n",
      "63\t12\t0.0\t1\t11 1\n",
      "64\t4\t0.0\t1\t4\n",
      "65\t43\t0.0\t1\t43\n",
      "66\t66\t0.0\t1\t66\n",
      "67\t14\t0.0\t1\t14\n",
      "68\t35\t0.0\t1\t33 2\n",
      "69\t50\t0.0\t1\t49 1\n",
      "70\t88\t0.0\t1\t87 1\n",
      "71\t39\t0.0\t1\t39\n",
      "72\t6\t0.0\t1\t6\n",
      "73\t8\t0.0\t1\t6 2\n",
      "74\t22\t0.0\t1\t21 1\n",
      "75\t41\t0.0\t1\t40 1\n",
      "76\t51\t0.0\t1\t51\n",
      "77\t50\t0.0\t1\t49 1\n",
      "78\t42\t0.0\t1\t42\n",
      "79\t47\t0.0\t1\t45 2\n",
      "80\t49\t0.0\t1\t48 1\n",
      "81\t51\t0.0\t1\t50 1\n",
      "82\t55\t0.0\t1\t54 1\n",
      "83\t58\t0.0\t1\t54 4\n",
      "84\t66\t0.0\t1\t65 1\n",
      "85\t48\t0.0\t1\t46 2\n",
      "86\t64\t0.0\t1\t60 4\n",
      "87\t51\t0.0\t1\t50 1\n",
      "88\t58\t0.0\t1\t57 1\n",
      "89\t60\t0.0\t1\t60\n",
      "90\t56\t0.0\t1\t56\n",
      "91\t57\t0.0\t1\t57\n",
      "92\t51\t0.0\t1\t51\n",
      "93\t39\t0.0\t1\t39\n",
      "94\t56\t0.0\t1\t54 2\n",
      "95\t58\t0.0\t1\t57 1\n",
      "96\t71\t0.0\t1\t69 2\n",
      "97\t68\t0.0\t1\t65 3\n",
      "98\t77\t0.0\t1\t75 2\n",
      "99\t48\t0.0\t1\t48\n",
      "100\t56\t0.0\t1\t55 1\n",
      "101\t52\t0.0\t1\t52\n",
      "102\t49\t0.0\t1\t49\n",
      "103\t60\t0.0\t1\t59 1\n",
      "104\t76\t0.0\t1\t76\n",
      "105\t62\t0.0\t1\t60 2\n",
      "106\t74\t0.0\t1\t74\n",
      "107\t58\t0.0\t1\t58\n",
      "108\t68\t0.0\t1\t67 1\n",
      "109\t56\t0.0\t1\t56\n",
      "110\t50\t0.0\t1\t49 1\n",
      "111\t43\t0.0\t1\t42 1\n",
      "112\t34\t0.0\t1\t34\n",
      "113\t35\t0.0\t1\t33 2\n",
      "114\t36\t0.0\t1\t33 3\n",
      "115\t15\t0.0\t1\t15\n",
      "116\t15\t0.0\t1\t15\n",
      "117\t7\t0.0\t1\t6 1\n",
      "118\t4\t0.0\t1\t4\n",
      "119\t2\t0.0\t1\t2\n",
      "121\t2\t0.0\t1\t2\n",
      "122\t6\t0.0\t1\t5 1\n",
      "129\t1\t0.0\t1\t0 1\n",
      "137\t2\t0.0\t1\t2\n",
      "139\t1\t0.0\t1\t1\n",
      "149\t1\t0.0\t1\t0 1\n",
      "150\t4\t0.0\t1\t4\n",
      "\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /scratch/hansetj1/script_testing/fastq_processing/raw_reads/sub_read1.fq\n",
      "=============================================\n",
      "10000 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Writing report to '/gpfs23/scratch/hansetj1/script_testing/fastq_processing/trimmed_reads/sub_read2.fq_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /scratch/hansetj1/script_testing/fastq_processing/raw_reads/sub_read2.fq\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.6\n",
      "Cutadapt version: 1.18\n",
      "Number of cores used for trimming: 1\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'CTGTCTCTTATA' (Nextera Transposase sequence; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Running FastQC on the data once trimming has completed\n",
      "Running FastQC with the following extra arguments: '--outdir /scratch/hansetj1/script_testing/fastq_processing/qc'\n",
      "\n",
      "Cutadapt seems to be reasonably up-to-date. Setting -j -j 1\n",
      "Writing final adapter and quality trimmed output to sub_read2_trimmed.fq\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'CTGTCTCTTATA' from file /scratch/hansetj1/script_testing/fastq_processing/raw_reads/sub_read2.fq <<< \n",
      "This is cutadapt 1.18 with Python 3.6.13\n",
      "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a CTGTCTCTTATA /scratch/hansetj1/script_testing/fastq_processing/raw_reads/sub_read2.fq\n",
      "Processing reads on 1 core in single-end mode ...\n",
      "Finished in 0.95 s (95 us/read; 0.63 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:                  10,000\n",
      "Reads with adapters:                     7,330 (73.3%)\n",
      "Reads written (passing filters):        10,000 (100.0%)\n",
      "\n",
      "Total basepairs processed:     1,496,688 bp\n",
      "Quality-trimmed:                   6,421 bp (0.4%)\n",
      "Total written (filtered):      1,156,722 bp (77.3%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: CTGTCTCTTATA; Type: regular 3'; Length: 12; Trimmed: 7330 times.\n",
      "\n",
      "No. of allowed errors:\n",
      "0-9 bp: 0; 10-12 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 13.5%\n",
      "  C: 40.1%\n",
      "  G: 26.4%\n",
      "  T: 19.9%\n",
      "  none/other: 0.1%\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t1652\t2500.0\t0\t1652\n",
      "2\t449\t625.0\t0\t449\n",
      "3\t152\t156.2\t0\t152\n",
      "4\t55\t39.1\t0\t55\n",
      "5\t41\t9.8\t0\t41\n",
      "6\t32\t2.4\t0\t32\n",
      "7\t26\t0.6\t0\t26\n",
      "8\t31\t0.2\t0\t31\n",
      "9\t43\t0.0\t0\t43\n",
      "11\t50\t0.0\t1\t48 2\n",
      "13\t57\t0.0\t1\t55 2\n",
      "14\t1\t0.0\t1\t1\n",
      "15\t79\t0.0\t1\t79\n",
      "16\t2\t0.0\t1\t2\n",
      "17\t26\t0.0\t1\t25 1\n",
      "18\t30\t0.0\t1\t27 3\n",
      "19\t29\t0.0\t1\t29\n",
      "20\t62\t0.0\t1\t60 2\n",
      "21\t1\t0.0\t1\t0 1\n",
      "22\t30\t0.0\t1\t28 2\n",
      "23\t27\t0.0\t1\t26 1\n",
      "24\t31\t0.0\t1\t31\n",
      "25\t43\t0.0\t1\t42 1\n",
      "26\t28\t0.0\t1\t27 1\n",
      "27\t37\t0.0\t1\t37\n",
      "28\t39\t0.0\t1\t39\n",
      "29\t62\t0.0\t1\t61 1\n",
      "30\t1\t0.0\t1\t0 1\n",
      "31\t31\t0.0\t1\t31\n",
      "32\t75\t0.0\t1\t73 2\n",
      "33\t2\t0.0\t1\t1 1\n",
      "34\t39\t0.0\t1\t38 1\n",
      "35\t38\t0.0\t1\t38\n",
      "36\t30\t0.0\t1\t30\n",
      "37\t36\t0.0\t1\t33 3\n",
      "38\t38\t0.0\t1\t36 2\n",
      "39\t62\t0.0\t1\t60 2\n",
      "40\t5\t0.0\t1\t4 1\n",
      "41\t48\t0.0\t1\t48\n",
      "42\t42\t0.0\t1\t39 3\n",
      "43\t39\t0.0\t1\t39\n",
      "44\t67\t0.0\t1\t65 2\n",
      "45\t205\t0.0\t1\t200 5\n",
      "46\t3\t0.0\t1\t3\n",
      "47\t16\t0.0\t1\t16\n",
      "48\t2\t0.0\t1\t2\n",
      "49\t18\t0.0\t1\t17 1\n",
      "50\t4\t0.0\t1\t4\n",
      "51\t39\t0.0\t1\t39\n",
      "52\t23\t0.0\t1\t23\n",
      "53\t58\t0.0\t1\t58\n",
      "54\t61\t0.0\t1\t61\n",
      "55\t28\t0.0\t1\t28\n",
      "56\t48\t0.0\t1\t46 2\n",
      "57\t140\t0.0\t1\t139 1\n",
      "58\t35\t0.0\t1\t34 1\n",
      "59\t22\t0.0\t1\t20 2\n",
      "60\t99\t0.0\t1\t98 1\n",
      "61\t29\t0.0\t1\t29\n",
      "62\t35\t0.0\t1\t35\n",
      "63\t167\t0.0\t1\t166 1\n",
      "64\t103\t0.0\t1\t102 1\n",
      "65\t11\t0.0\t1\t11\n",
      "66\t52\t0.0\t1\t50 2\n",
      "67\t71\t0.0\t1\t70 1\n",
      "68\t12\t0.0\t1\t11 1\n",
      "69\t23\t0.0\t1\t22 1\n",
      "70\t74\t0.0\t1\t74\n",
      "71\t32\t0.0\t1\t32\n",
      "72\t24\t0.0\t1\t23 1\n",
      "73\t36\t0.0\t1\t33 3\n",
      "74\t33\t0.0\t1\t33\n",
      "75\t44\t0.0\t1\t44\n",
      "76\t46\t0.0\t1\t44 2\n",
      "77\t40\t0.0\t1\t39 1\n",
      "78\t41\t0.0\t1\t41\n",
      "79\t41\t0.0\t1\t41\n",
      "80\t39\t0.0\t1\t39\n",
      "81\t47\t0.0\t1\t47\n",
      "82\t50\t0.0\t1\t50\n",
      "83\t60\t0.0\t1\t58 2\n",
      "84\t59\t0.0\t1\t58 1\n",
      "85\t56\t0.0\t1\t55 1\n",
      "86\t59\t0.0\t1\t58 1\n",
      "87\t50\t0.0\t1\t50\n",
      "88\t58\t0.0\t1\t58\n",
      "89\t53\t0.0\t1\t53\n",
      "90\t55\t0.0\t1\t55\n",
      "91\t57\t0.0\t1\t57\n",
      "92\t54\t0.0\t1\t54\n",
      "93\t43\t0.0\t1\t43\n",
      "94\t50\t0.0\t1\t49 1\n",
      "95\t58\t0.0\t1\t57 1\n",
      "96\t69\t0.0\t1\t69\n",
      "97\t59\t0.0\t1\t59\n",
      "98\t75\t0.0\t1\t75\n",
      "99\t55\t0.0\t1\t55\n",
      "100\t57\t0.0\t1\t54 3\n",
      "101\t54\t0.0\t1\t53 1\n",
      "102\t43\t0.0\t1\t43\n",
      "103\t55\t0.0\t1\t54 1\n",
      "104\t70\t0.0\t1\t69 1\n",
      "105\t60\t0.0\t1\t60\n",
      "106\t74\t0.0\t1\t74\n",
      "107\t53\t0.0\t1\t53\n",
      "108\t67\t0.0\t1\t66 1\n",
      "109\t53\t0.0\t1\t53\n",
      "110\t48\t0.0\t1\t48\n",
      "111\t41\t0.0\t1\t38 3\n",
      "112\t35\t0.0\t1\t35\n",
      "113\t36\t0.0\t1\t35 1\n",
      "114\t34\t0.0\t1\t34\n",
      "115\t14\t0.0\t1\t14\n",
      "116\t14\t0.0\t1\t14\n",
      "117\t5\t0.0\t1\t5\n",
      "118\t4\t0.0\t1\t4\n",
      "119\t2\t0.0\t1\t2\n",
      "121\t2\t0.0\t1\t1 1\n",
      "122\t7\t0.0\t1\t5 2\n",
      "125\t1\t0.0\t1\t0 1\n",
      "126\t1\t0.0\t1\t0 1\n",
      "132\t1\t0.0\t1\t0 1\n",
      "135\t1\t0.0\t1\t0 1\n",
      "137\t2\t0.0\t1\t2\n",
      "139\t1\t0.0\t1\t1\n",
      "147\t1\t0.0\t1\t0 1\n",
      "149\t1\t0.0\t1\t0 1\n",
      "150\t4\t0.0\t1\t4\n",
      "\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /scratch/hansetj1/script_testing/fastq_processing/raw_reads/sub_read2.fq\n",
      "=============================================\n",
      "10000 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files sub_read1_trimmed.fq and sub_read2_trimmed.fq\n",
      "Now renaming the output files\n",
      "\n",
      "ORIGINAL FILE 1: >>sub_read1_trimmed.fq<<\tRENAMING TO:>>test_set_2_R1_trimmed.fq<<\n",
      "ORIGINAL FILE 2: >>sub_read2_trimmed.fq<<\tRENAMING TO:>>test_set_2_R2_trimmed.fq<<\n",
      "file_1: test_set_2_R1_trimmed.fq, file_2: test_set_2_R2_trimmed.fq\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: test_set_2_R1_trimmed.fq and test_set_2_R2_trimmed.fq <<<<<\n",
      "Renaming the output files (AGAIN).\n",
      "Writing validated paired-end Read 1 reads to test_set_2_val_1.fq\n",
      "Writing validated paired-end Read 2 reads to test_set_2_val_2.fq\n",
      "\n",
      "Total number of sequences analysed: 10000\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 12 (0.12%)\n",
      "\n",
      "\n",
      "  >>> Now running FastQC on the validated data test_set_2_val_1.fq<<<\n",
      "\n",
      "Started analysis of test_set_2_val_1.fq\n",
      "Approx 10% complete for test_set_2_val_1.fq\n",
      "Approx 20% complete for test_set_2_val_1.fq\n",
      "Approx 30% complete for test_set_2_val_1.fq\n",
      "Approx 40% complete for test_set_2_val_1.fq\n",
      "Approx 50% complete for test_set_2_val_1.fq\n",
      "Approx 60% complete for test_set_2_val_1.fq\n",
      "Approx 70% complete for test_set_2_val_1.fq\n",
      "Approx 80% complete for test_set_2_val_1.fq\n",
      "Approx 90% complete for test_set_2_val_1.fq\n",
      "Analysis complete for test_set_2_val_1.fq\n",
      "\n",
      "  >>> Now running FastQC on the validated data test_set_2_val_2.fq<<<\n",
      "\n",
      "Started analysis of test_set_2_val_2.fq\n",
      "Approx 10% complete for test_set_2_val_2.fq\n",
      "Approx 20% complete for test_set_2_val_2.fq\n",
      "Approx 30% complete for test_set_2_val_2.fq\n",
      "Approx 40% complete for test_set_2_val_2.fq\n",
      "Approx 50% complete for test_set_2_val_2.fq\n",
      "Approx 60% complete for test_set_2_val_2.fq\n",
      "Approx 70% complete for test_set_2_val_2.fq\n",
      "Approx 80% complete for test_set_2_val_2.fq\n",
      "Approx 90% complete for test_set_2_val_2.fq\n",
      "Analysis complete for test_set_2_val_2.fq\n",
      "Deleting both intermediate output files test_set_2_R1_trimmed.fq and test_set_2_R2_trimmed.fq\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "\n",
      "***Mapping trimmed reads\n",
      "9988 reads; of these:\n",
      "  9988 (100.00%) were paired; of these:\n",
      "    721 (7.22%) aligned concordantly 0 times\n",
      "    4821 (48.27%) aligned concordantly exactly 1 time\n",
      "    4446 (44.51%) aligned concordantly >1 times\n",
      "92.78% overall alignment rate\n",
      "\n",
      "***Filtering mapped reads\n",
      "[bam_sort_core] merging from 0 files and 8 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 8 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 8 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 8 in-memory blocks...\n",
      "INFO\t2021-10-19 14:24:16\tMarkDuplicates\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    MarkDuplicates -I /scratch/hansetj1/script_testing/fastq_processing/bams/test_set_2.filtered.pos-sorted.bam -O /scratch/hansetj1/script_testing/fastq_processing/bams/test_set_2.unique.unsorted.bam -M /scratch/hansetj1/script_testing/fastq_processing/qc/test_set_2_marked_dup_metrics-all.txt -REMOVE_DUPLICATES TRUE\n",
      "**********\n",
      "\n",
      "\n",
      "14:24:17.247 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/gpfs52/home/hansetj1/.conda/envs/jupyter/share/picard-2.26.3-0/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Tue Oct 19 14:24:17 CDT 2021] MarkDuplicates INPUT=[/scratch/hansetj1/script_testing/fastq_processing/bams/test_set_2.filtered.pos-sorted.bam] OUTPUT=/scratch/hansetj1/script_testing/fastq_processing/bams/test_set_2.unique.unsorted.bam METRICS_FILE=/scratch/hansetj1/script_testing/fastq_processing/qc/test_set_2_marked_dup_metrics-all.txt REMOVE_DUPLICATES=true    MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP=50000 MAX_FILE_HANDLES_FOR_READ_ENDS_MAP=8000 SORTING_COLLECTION_SIZE_RATIO=0.25 TAG_DUPLICATE_SET_MEMBERS=false REMOVE_SEQUENCING_DUPLICATES=false TAGGING_POLICY=DontTag CLEAR_DT=true DUPLEX_UMI=false ADD_PG_TAG_TO_READS=true ASSUME_SORTED=false DUPLICATE_SCORING_STRATEGY=SUM_OF_BASE_QUALITIES PROGRAM_RECORD_ID=MarkDuplicates PROGRAM_GROUP_NAME=MarkDuplicates READ_NAME_REGEX=<optimized capture of last three ':' separated fields as numeric values> OPTICAL_DUPLICATE_PIXEL_DISTANCE=100 MAX_OPTICAL_DUPLICATE_SET_SIZE=300000 VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false GA4GH_CLIENT_SECRETS=client_secrets.json USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
      "[Tue Oct 19 14:24:17 CDT 2021] Executing as hansetj1@cn1288.vampire on Linux 3.10.0-1160.36.2.el7.x86_64 amd64; OpenJDK 64-Bit Server VM 1.8.0_302-b08; Deflater: Intel; Inflater: Intel; Provider GCS is not available; Picard version: 2.26.3\n",
      "INFO\t2021-10-19 14:24:17\tMarkDuplicates\tStart of doWork freeMemory: 1117100088; totalMemory: 1133510656; maxMemory: 16799760384\n",
      "INFO\t2021-10-19 14:24:17\tMarkDuplicates\tReading input file and constructing read end information.\n",
      "INFO\t2021-10-19 14:24:17\tMarkDuplicates\tWill retain up to 60868697 data points before spilling to disk.\n",
      "INFO\t2021-10-19 14:24:17\tMarkDuplicates\tRead 7464 records. 0 pairs never matched.\n",
      "INFO\t2021-10-19 14:24:18\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 635141376; totalMemory: 1133510656; maxMemory: 16799760384\n",
      "INFO\t2021-10-19 14:24:18\tMarkDuplicates\tWill retain up to 524992512 duplicate indices before spilling to disk.\n",
      "INFO\t2021-10-19 14:24:21\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
      "INFO\t2021-10-19 14:24:21\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
      "INFO\t2021-10-19 14:24:21\tMarkDuplicates\tSorting list of duplicate records.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO\t2021-10-19 14:24:22\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 1123623056; totalMemory: 5333581824; maxMemory: 16799760384\n",
      "INFO\t2021-10-19 14:24:22\tMarkDuplicates\tMarking 0 records as duplicates.\n",
      "INFO\t2021-10-19 14:24:22\tMarkDuplicates\tFound 0 optical duplicate clusters.\n",
      "INFO\t2021-10-19 14:24:22\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
      "INFO\t2021-10-19 14:24:23\tMarkDuplicates\tWriting complete. Closing input iterator.\n",
      "INFO\t2021-10-19 14:24:23\tMarkDuplicates\tDuplicate Index cleanup.\n",
      "INFO\t2021-10-19 14:24:23\tMarkDuplicates\tGetting Memory Stats.\n",
      "INFO\t2021-10-19 14:24:23\tMarkDuplicates\tBefore output close freeMemory: 5322923168; totalMemory: 5333581824; maxMemory: 16799760384\n",
      "INFO\t2021-10-19 14:24:23\tMarkDuplicates\tClosed outputs. Getting more Memory Stats.\n",
      "INFO\t2021-10-19 14:24:23\tMarkDuplicates\tAfter output close freeMemory: 5323004456; totalMemory: 5333581824; maxMemory: 16799760384\n",
      "[Tue Oct 19 14:24:24 CDT 2021] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 0.13 minutes.\n",
      "Runtime.totalMemory()=5333581824\n",
      "[bam_sort_core] merging from 0 files and 8 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 8 in-memory blocks...\n",
      "\n",
      "***Counting mapped reads\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "dir='/scratch/hansetj1/script_testing/fastq_processing'\n",
    "\n",
    "python3 ./fastq_processing/fastq_processing.py -a ${dir}/raw_reads/sub_read1.fq -b ${dir}/raw_reads/sub_read2.fq \\\n",
    "    -n test_set_2 -e /home/hansetj1/hg38_encode_blacklist_ENCFF356LFX.bed -t ${dir}/trimmed_reads \\\n",
    "    -o ${dir}/bams -r ${dir}/qc -g /data/hodges_lab/hg38_genome/bt2/hg38"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "669e18e7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 5.8M\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 2.9M Oct 19 14:22 test_set_2_val_1.fq\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 2.9M Oct 19 14:22 test_set_2_val_2.fq\n",
      "total 19M\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 7.2M Oct 19 14:24 test_set_2.sam\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 817K Oct 19 14:24 test_set_2.bam\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 1.5M Oct 19 14:24 test_set_2.bam.bai\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 538K Oct 19 14:24 test_set_2.no_chrM.unsorted.bam\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 538K Oct 19 14:24 test_set_2.no_chrM.bam\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 1.5M Oct 19 14:24 test_set_2.no_chrM.bam.bai\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 7.9K Oct 19 14:24 test_set_2.blacklisted.bam\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 537K Oct 19 14:24 test_set_2.filtered.unsorted.bam\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 537K Oct 19 14:24 test_set_2.filtered.pos-sorted.bam\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 1.5M Oct 19 14:24 test_set_2.filtered.pos-sorted.bam.bai\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 542K Oct 19 14:24 test_set_2.filtered.n-sorted.bam\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 540K Oct 19 14:24 test_set_2.unique.unsorted.bam\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 541K Oct 19 14:24 test_set_2.unique.pos-sorted.bam\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 1.5M Oct 19 14:24 test_set_2.unique.pos-sorted.bam.bai\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 546K Oct 19 14:24 test_set_2.unique.n-sorted.bam\n",
      "total 1.4M\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 3.9K Oct 19 14:22 sub_read1.fq_trimming_report.txt\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 4.1K Oct 19 14:22 sub_read2.fq_trimming_report.txt\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 372K Oct 19 14:23 test_set_2_val_1_fastqc.zip\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 262K Oct 19 14:23 test_set_2_val_1_fastqc.html\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 376K Oct 19 14:23 test_set_2_val_2_fastqc.zip\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 261K Oct 19 14:23 test_set_2_val_2_fastqc.html\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab 1.6K Oct 19 14:24 test_set_2_marked_dup_metrics-all.txt\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab  488 Oct 19 14:24 test_set_2.sam_flagstat-results.txt\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab  491 Oct 19 14:24 test_set_2.bam_flagstat-results.txt\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab  484 Oct 19 14:24 test_set_2.no_chrM.bam_flagstat-results.txt\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab  484 Oct 19 14:24 test_set_2.filtered.pos-sorted.bam_flagstat-results.txt\n",
      "-rw-r--r-- 1 hansetj1 hodges_lab  484 Oct 19 14:24 test_set_2.unique.pos-sorted.bam_flagstat-results.txt\n"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "#check output files\n",
    "\n",
    "dir='/scratch/hansetj1/script_testing/fastq_processing'\n",
    "\n",
    "ls -lhtr ${dir}/trimmed_reads\n",
    "ls -lhtr ${dir}/bams\n",
    "ls -lhtr ${dir}/qc\n",
    "\n",
    "samtools view ${dir}/bams/test_set_2.no_chrM.bam | grep chrM "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "5e8d6cc2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19976 + 0 in total (QC-passed reads + QC-failed reads)\n",
      "19976 + 0 primary\n",
      "0 + 0 secondary\n",
      "0 + 0 supplementary\n",
      "0 + 0 duplicates\n",
      "0 + 0 primary duplicates\n",
      "18534 + 0 mapped (92.78% : N/A)\n",
      "18534 + 0 primary mapped (92.78% : N/A)\n",
      "19976 + 0 paired in sequencing\n",
      "9988 + 0 read1\n",
      "9988 + 0 read2\n",
      "18534 + 0 properly paired (92.78% : N/A)\n",
      "18534 + 0 with itself and mate mapped\n",
      "0 + 0 singletons (0.00% : N/A)\n",
      "0 + 0 with mate mapped to a different chr\n",
      "0 + 0 with mate mapped to a different chr (mapQ>=5)\n",
      "14574 + 0 in total (QC-passed reads + QC-failed reads)\n",
      "14574 + 0 primary\n",
      "0 + 0 secondary\n",
      "0 + 0 supplementary\n",
      "0 + 0 duplicates\n",
      "0 + 0 primary duplicates\n",
      "14574 + 0 mapped (100.00% : N/A)\n",
      "14574 + 0 primary mapped (100.00% : N/A)\n",
      "14574 + 0 paired in sequencing\n",
      "7287 + 0 read1\n",
      "7287 + 0 read2\n",
      "14574 + 0 properly paired (100.00% : N/A)\n",
      "14574 + 0 with itself and mate mapped\n",
      "0 + 0 singletons (0.00% : N/A)\n",
      "0 + 0 with mate mapped to a different chr\n",
      "0 + 0 with mate mapped to a different chr (mapQ>=5)\n",
      "7484 + 0 in total (QC-passed reads + QC-failed reads)\n",
      "7484 + 0 primary\n",
      "0 + 0 secondary\n",
      "0 + 0 supplementary\n",
      "0 + 0 duplicates\n",
      "0 + 0 primary duplicates\n",
      "7484 + 0 mapped (100.00% : N/A)\n",
      "7484 + 0 primary mapped (100.00% : N/A)\n",
      "7484 + 0 paired in sequencing\n",
      "3742 + 0 read1\n",
      "3742 + 0 read2\n",
      "7484 + 0 properly paired (100.00% : N/A)\n",
      "7484 + 0 with itself and mate mapped\n",
      "0 + 0 singletons (0.00% : N/A)\n",
      "0 + 0 with mate mapped to a different chr\n",
      "0 + 0 with mate mapped to a different chr (mapQ>=5)\n",
      "7464 + 0 in total (QC-passed reads + QC-failed reads)\n",
      "7464 + 0 primary\n",
      "0 + 0 secondary\n",
      "0 + 0 supplementary\n",
      "0 + 0 duplicates\n",
      "0 + 0 primary duplicates\n",
      "7464 + 0 mapped (100.00% : N/A)\n",
      "7464 + 0 primary mapped (100.00% : N/A)\n",
      "7464 + 0 paired in sequencing\n",
      "3732 + 0 read1\n",
      "3732 + 0 read2\n",
      "7464 + 0 properly paired (100.00% : N/A)\n",
      "7464 + 0 with itself and mate mapped\n",
      "0 + 0 singletons (0.00% : N/A)\n",
      "0 + 0 with mate mapped to a different chr\n",
      "0 + 0 with mate mapped to a different chr (mapQ>=5)\n",
      "7464 + 0 in total (QC-passed reads + QC-failed reads)\n",
      "7464 + 0 primary\n",
      "0 + 0 secondary\n",
      "0 + 0 supplementary\n",
      "0 + 0 duplicates\n",
      "0 + 0 primary duplicates\n",
      "7464 + 0 mapped (100.00% : N/A)\n",
      "7464 + 0 primary mapped (100.00% : N/A)\n",
      "7464 + 0 paired in sequencing\n",
      "3732 + 0 read1\n",
      "3732 + 0 read2\n",
      "7464 + 0 properly paired (100.00% : N/A)\n",
      "7464 + 0 with itself and mate mapped\n",
      "0 + 0 singletons (0.00% : N/A)\n",
      "0 + 0 with mate mapped to a different chr\n",
      "0 + 0 with mate mapped to a different chr (mapQ>=5)\n"
     ]
    }
   ],
   "source": [
    "#The chrM filter seems to have worked. Check the flagstat results to see how read count was lost. \n",
    "cat ${dir}/qc/test_set_2.sam_flagstat-results.txt ${dir}/qc/test_set_2.bam_flagstat-results.txt \\\n",
    "    ${dir}/qc/test_set_2.no_chrM.bam_flagstat-results.txt ${dir}/qc/test_set_2.filtered.pos-sorted.bam_flagstat-results.txt \\\n",
    "    ${dir}/qc/test_set_2.unique.pos-sorted.bam_flagstat-results.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "df50b747",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7464 + 0 in total (QC-passed reads + QC-failed reads)\n",
      "7464 + 0 primary\n",
      "0 + 0 secondary\n",
      "0 + 0 supplementary\n",
      "0 + 0 duplicates\n",
      "0 + 0 primary duplicates\n",
      "7464 + 0 mapped (100.00% : N/A)\n",
      "7464 + 0 primary mapped (100.00% : N/A)\n",
      "7464 + 0 paired in sequencing\n",
      "3732 + 0 read1\n",
      "3732 + 0 read2\n",
      "7464 + 0 properly paired (100.00% : N/A)\n",
      "7464 + 0 with itself and mate mapped\n",
      "0 + 0 singletons (0.00% : N/A)\n",
      "0 + 0 with mate mapped to a different chr\n",
      "0 + 0 with mate mapped to a different chr (mapQ>=5)\n",
      "7464 + 0 in total (QC-passed reads + QC-failed reads)\n",
      "7464 + 0 primary\n",
      "0 + 0 secondary\n",
      "0 + 0 supplementary\n",
      "0 + 0 duplicates\n",
      "0 + 0 primary duplicates\n",
      "7464 + 0 mapped (100.00% : N/A)\n",
      "7464 + 0 primary mapped (100.00% : N/A)\n",
      "7464 + 0 paired in sequencing\n",
      "3732 + 0 read1\n",
      "3732 + 0 read2\n",
      "7464 + 0 properly paired (100.00% : N/A)\n",
      "7464 + 0 with itself and mate mapped\n",
      "0 + 0 singletons (0.00% : N/A)\n",
      "0 + 0 with mate mapped to a different chr\n",
      "0 + 0 with mate mapped to a different chr (mapQ>=5)\n"
     ]
    }
   ],
   "source": [
    "#There seems to be an issue with the duplicate removal as no reads were lost. \n",
    "cat ${dir}/qc/test_set_2.filtered.pos-sorted.bam_flagstat-results.txt \\\n",
    "    ${dir}/qc/test_set_2.unique.pos-sorted.bam_flagstat-results.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f60de343",
   "metadata": {},
   "source": [
    "### It is possible that there are no duplicates in this file. \n",
    "\n",
    "I will try an RNA file this time, so we should have more duplicates. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "e2c71734",
   "metadata": {},
   "outputs": [],
   "source": [
    "#move fastq files to new folder\n",
    "\n",
    "dir='/data/hodges_lab/ATAC-STARR/data/1_FASTQ'\n",
    "\n",
    "gunzip -c ${dir}/3108-TH-6-TAGGCATG-GTAAGGAG_S406_R1_001.fastq.gz > /scratch/hansetj1/script_testing/fastq_processing/raw_reads/RNA_read1.fq\n",
    "gunzip -c ${dir}/3108-TH-6-TAGGCATG-GTAAGGAG_S406_R2_001.fastq.gz > /scratch/hansetj1/script_testing/fastq_processing/raw_reads/RNA_read2.fq\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "8c12672d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Subsample fastq files for quick processing using seqtk\n",
    "\n",
    "dir='/scratch/hansetj1/script_testing/fastq_processing/raw_reads'\n",
    "\n",
    "# FASTQ R1\n",
    "seqtk sample -s 123 ${dir}/RNA_read1.fq 10000 > ${dir}/RNA_sub_read1.fq\n",
    " \n",
    "# FASTQ R2\n",
    "seqtk sample -s 123 ${dir}/RNA_read2.fq 10000 > ${dir}/RNA_sub_read2.fq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "61f1c442",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PARAMETERS:\n",
      "Fastq R1: /scratch/hansetj1/script_testing/fastq_processing/raw_reads/RNA_sub_read1.fq\n",
      "Fastq R2: /scratch/hansetj1/script_testing/fastq_processing/raw_reads/RNA_sub_read2.fq\n",
      "Sample name: test_set_RNA\n",
      "Blacklist Bedfile /home/hansetj1/hg38_encode_blacklist_ENCFF356LFX.bed\n",
      "Trimmed reads directory: /scratch/hansetj1/script_testing/fastq_processing/trimmed_reads\n",
      "Mapped reads directory: /scratch/hansetj1/script_testing/fastq_processing/bams\n",
      "QC results directory: /scratch/hansetj1/script_testing/fastq_processing/qc\n",
      "Genome: /data/hodges_lab/hg38_genome/bt2/hg38\n",
      "Detected and using 8 cores for processing\n",
      "\n",
      "***Trimming raw reads\n",
      "Multicore support not enabled. Proceeding with single-core trimming.\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 1.18\n",
      "single-core operation.\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output will be written into the directory: /gpfs23/scratch/hansetj1/script_testing/fastq_processing/trimmed_reads/\n",
      "Using user-specified basename (>>test_set_RNA<<) instead of deriving the filename from the input file(s)\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> /scratch/hansetj1/script_testing/fastq_processing/raw_reads/RNA_sub_read1.fq <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "Nextera\t6469\tCTGTCTCTTATA\t10000\t64.69\n",
      "smallRNA\t0\tTGGAATTCTCGG\t10000\t0.00\n",
      "Illumina\t0\tAGATCGGAAGAGC\t10000\t0.00\n",
      "Using Nextera adapter for trimming (count: 6469). Second best hit was smallRNA (count: 0)\n",
      "\n",
      "Writing report to '/gpfs23/scratch/hansetj1/script_testing/fastq_processing/trimmed_reads/RNA_sub_read1.fq_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /scratch/hansetj1/script_testing/fastq_processing/raw_reads/RNA_sub_read1.fq\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.6\n",
      "Cutadapt version: 1.18\n",
      "Number of cores used for trimming: 1\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'CTGTCTCTTATA' (Nextera Transposase sequence; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Running FastQC on the data once trimming has completed\n",
      "Running FastQC with the following extra arguments: '--outdir /scratch/hansetj1/script_testing/fastq_processing/qc'\n",
      "\n",
      "Cutadapt seems to be reasonably up-to-date. Setting -j 1\n",
      "Writing final adapter and quality trimmed output to RNA_sub_read1_trimmed.fq\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'CTGTCTCTTATA' from file /scratch/hansetj1/script_testing/fastq_processing/raw_reads/RNA_sub_read1.fq <<< \n",
      "This is cutadapt 1.18 with Python 3.6.13\n",
      "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a CTGTCTCTTATA /scratch/hansetj1/script_testing/fastq_processing/raw_reads/RNA_sub_read1.fq\n",
      "Processing reads on 1 core in single-end mode ...\n",
      "Finished in 0.16 s (16 us/read; 3.83 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:                  10,000\n",
      "Reads with adapters:                     8,371 (83.7%)\n",
      "Reads written (passing filters):        10,000 (100.0%)\n",
      "\n",
      "Total basepairs processed:     1,497,325 bp\n",
      "Quality-trimmed:                   1,831 bp (0.1%)\n",
      "Total written (filtered):        999,989 bp (66.8%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: CTGTCTCTTATA; Type: regular 3'; Length: 12; Trimmed: 8371 times.\n",
      "\n",
      "No. of allowed errors:\n",
      "0-9 bp: 0; 10-12 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 11.9%\n",
      "  C: 42.4%\n",
      "  G: 25.9%\n",
      "  T: 19.7%\n",
      "  none/other: 0.1%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t1131\t2500.0\t0\t1131\n",
      "2\t290\t625.0\t0\t290\n",
      "3\t115\t156.2\t0\t115\n",
      "4\t37\t39.1\t0\t37\n",
      "5\t31\t9.8\t0\t31\n",
      "6\t24\t2.4\t0\t24\n",
      "7\t29\t0.6\t0\t29\n",
      "8\t30\t0.2\t0\t30\n",
      "9\t38\t0.0\t0\t38\n",
      "10\t3\t0.0\t1\t3\n",
      "11\t53\t0.0\t1\t52 1\n",
      "12\t3\t0.0\t1\t2 1\n",
      "13\t60\t0.0\t1\t60\n",
      "14\t1\t0.0\t1\t0 1\n",
      "15\t63\t0.0\t1\t63\n",
      "17\t23\t0.0\t1\t23\n",
      "18\t34\t0.0\t1\t34\n",
      "19\t30\t0.0\t1\t30\n",
      "20\t31\t0.0\t1\t30 1\n",
      "21\t31\t0.0\t1\t29 2\n",
      "22\t103\t0.0\t1\t102 1\n",
      "23\t1\t0.0\t1\t0 1\n",
      "24\t3\t0.0\t1\t2 1\n",
      "25\t23\t0.0\t1\t23\n",
      "26\t36\t0.0\t1\t35 1\n",
      "27\t93\t0.0\t1\t92 1\n",
      "28\t1\t0.0\t1\t0 1\n",
      "29\t32\t0.0\t1\t32\n",
      "30\t151\t0.0\t1\t149 2\n",
      "31\t3\t0.0\t1\t2 1\n",
      "34\t35\t0.0\t1\t33 2\n",
      "35\t118\t0.0\t1\t118\n",
      "37\t2\t0.0\t1\t1 1\n",
      "38\t41\t0.0\t1\t40 1\n",
      "39\t87\t0.0\t1\t85 2\n",
      "40\t3\t0.0\t1\t2 1\n",
      "41\t40\t0.0\t1\t39 1\n",
      "42\t100\t0.0\t1\t98 2\n",
      "43\t3\t0.0\t1\t2 1\n",
      "44\t46\t0.0\t1\t43 3\n",
      "45\t76\t0.0\t1\t76\n",
      "46\t23\t0.0\t1\t21 2\n",
      "47\t38\t0.0\t1\t38\n",
      "48\t52\t0.0\t1\t52\n",
      "49\t88\t0.0\t1\t88\n",
      "50\t3\t0.0\t1\t3\n",
      "51\t54\t0.0\t1\t52 2\n",
      "52\t57\t0.0\t1\t56 1\n",
      "53\t40\t0.0\t1\t40\n",
      "54\t52\t0.0\t1\t49 3\n",
      "55\t69\t0.0\t1\t69\n",
      "56\t54\t0.0\t1\t54\n",
      "57\t54\t0.0\t1\t52 2\n",
      "58\t65\t0.0\t1\t63 2\n",
      "59\t50\t0.0\t1\t48 2\n",
      "60\t54\t0.0\t1\t54\n",
      "61\t37\t0.0\t1\t35 2\n",
      "62\t82\t0.0\t1\t78 4\n",
      "63\t66\t0.0\t1\t64 2\n",
      "64\t55\t0.0\t1\t55\n",
      "65\t56\t0.0\t1\t55 1\n",
      "66\t123\t0.0\t1\t122 1\n",
      "67\t61\t0.0\t1\t59 2\n",
      "68\t88\t0.0\t1\t88\n",
      "69\t93\t0.0\t1\t92 1\n",
      "70\t103\t0.0\t1\t101 2\n",
      "71\t15\t0.0\t1\t14 1\n",
      "72\t11\t0.0\t1\t9 2\n",
      "73\t42\t0.0\t1\t41 1\n",
      "74\t47\t0.0\t1\t46 1\n",
      "75\t63\t0.0\t1\t62 1\n",
      "76\t77\t0.0\t1\t76 1\n",
      "77\t60\t0.0\t1\t59 1\n",
      "78\t69\t0.0\t1\t68 1\n",
      "79\t71\t0.0\t1\t71\n",
      "80\t65\t0.0\t1\t65\n",
      "81\t85\t0.0\t1\t85\n",
      "82\t65\t0.0\t1\t64 1\n",
      "83\t75\t0.0\t1\t74 1\n",
      "84\t90\t0.0\t1\t89 1\n",
      "85\t84\t0.0\t1\t83 1\n",
      "86\t107\t0.0\t1\t103 4\n",
      "87\t87\t0.0\t1\t85 2\n",
      "88\t98\t0.0\t1\t97 1\n",
      "89\t102\t0.0\t1\t97 5\n",
      "90\t101\t0.0\t1\t99 2\n",
      "91\t89\t0.0\t1\t86 3\n",
      "92\t93\t0.0\t1\t92 1\n",
      "93\t95\t0.0\t1\t94 1\n",
      "94\t88\t0.0\t1\t87 1\n",
      "95\t119\t0.0\t1\t119\n",
      "96\t106\t0.0\t1\t103 3\n",
      "97\t92\t0.0\t1\t90 2\n",
      "98\t93\t0.0\t1\t92 1\n",
      "99\t99\t0.0\t1\t97 2\n",
      "100\t101\t0.0\t1\t101\n",
      "101\t95\t0.0\t1\t94 1\n",
      "102\t92\t0.0\t1\t91 1\n",
      "103\t134\t0.0\t1\t131 3\n",
      "104\t104\t0.0\t1\t100 4\n",
      "105\t114\t0.0\t1\t111 3\n",
      "106\t110\t0.0\t1\t110\n",
      "107\t92\t0.0\t1\t92\n",
      "108\t104\t0.0\t1\t103 1\n",
      "109\t84\t0.0\t1\t83 1\n",
      "110\t84\t0.0\t1\t81 3\n",
      "111\t63\t0.0\t1\t63\n",
      "112\t70\t0.0\t1\t70\n",
      "113\t62\t0.0\t1\t61 1\n",
      "114\t54\t0.0\t1\t53 1\n",
      "115\t54\t0.0\t1\t54\n",
      "116\t26\t0.0\t1\t25 1\n",
      "117\t13\t0.0\t1\t12 1\n",
      "118\t9\t0.0\t1\t9\n",
      "119\t4\t0.0\t1\t4\n",
      "120\t6\t0.0\t1\t6\n",
      "121\t5\t0.0\t1\t5\n",
      "122\t4\t0.0\t1\t3 1\n",
      "123\t2\t0.0\t1\t2\n",
      "125\t2\t0.0\t1\t2\n",
      "127\t1\t0.0\t1\t1\n",
      "129\t1\t0.0\t1\t1\n",
      "130\t1\t0.0\t1\t1\n",
      "132\t1\t0.0\t1\t1\n",
      "135\t1\t0.0\t1\t1\n",
      "136\t3\t0.0\t1\t3\n",
      "137\t1\t0.0\t1\t1\n",
      "138\t1\t0.0\t1\t1\n",
      "140\t2\t0.0\t1\t1 1\n",
      "143\t1\t0.0\t1\t1\n",
      "148\t3\t0.0\t1\t3\n",
      "150\t8\t0.0\t1\t8\n",
      "\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /scratch/hansetj1/script_testing/fastq_processing/raw_reads/RNA_sub_read1.fq\n",
      "=============================================\n",
      "10000 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Writing report to '/gpfs23/scratch/hansetj1/script_testing/fastq_processing/trimmed_reads/RNA_sub_read2.fq_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /scratch/hansetj1/script_testing/fastq_processing/raw_reads/RNA_sub_read2.fq\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.6\n",
      "Cutadapt version: 1.18\n",
      "Number of cores used for trimming: 1\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'CTGTCTCTTATA' (Nextera Transposase sequence; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Running FastQC on the data once trimming has completed\n",
      "Running FastQC with the following extra arguments: '--outdir /scratch/hansetj1/script_testing/fastq_processing/qc'\n",
      "\n",
      "Cutadapt seems to be reasonably up-to-date. Setting -j -j 1\n",
      "Writing final adapter and quality trimmed output to RNA_sub_read2_trimmed.fq\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'CTGTCTCTTATA' from file /scratch/hansetj1/script_testing/fastq_processing/raw_reads/RNA_sub_read2.fq <<< \n",
      "This is cutadapt 1.18 with Python 3.6.13\n",
      "Command line parameters: -j 1 -e 0.1 -q 20 -O 1 -a CTGTCTCTTATA /scratch/hansetj1/script_testing/fastq_processing/raw_reads/RNA_sub_read2.fq\n",
      "Processing reads on 1 core in single-end mode ...\n",
      "Finished in 0.25 s (25 us/read; 2.42 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:                  10,000\n",
      "Reads with adapters:                     8,337 (83.4%)\n",
      "Reads written (passing filters):        10,000 (100.0%)\n",
      "\n",
      "Total basepairs processed:     1,497,126 bp\n",
      "Quality-trimmed:                   7,130 bp (0.5%)\n",
      "Total written (filtered):        999,287 bp (66.7%)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: CTGTCTCTTATA; Type: regular 3'; Length: 12; Trimmed: 8337 times.\n",
      "\n",
      "No. of allowed errors:\n",
      "0-9 bp: 0; 10-12 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 11.8%\n",
      "  C: 39.8%\n",
      "  G: 28.2%\n",
      "  T: 20.1%\n",
      "  none/other: 0.1%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t1074\t2500.0\t0\t1074\n",
      "2\t305\t625.0\t0\t305\n",
      "3\t113\t156.2\t0\t113\n",
      "4\t42\t39.1\t0\t42\n",
      "5\t27\t9.8\t0\t27\n",
      "6\t24\t2.4\t0\t24\n",
      "7\t27\t0.6\t0\t27\n",
      "8\t33\t0.2\t0\t33\n",
      "9\t39\t0.0\t0\t38 1\n",
      "10\t2\t0.0\t1\t2\n",
      "11\t54\t0.0\t1\t54\n",
      "12\t3\t0.0\t1\t2 1\n",
      "13\t62\t0.0\t1\t62\n",
      "15\t63\t0.0\t1\t63\n",
      "16\t2\t0.0\t1\t2\n",
      "17\t21\t0.0\t1\t21\n",
      "18\t34\t0.0\t1\t32 2\n",
      "19\t31\t0.0\t1\t30 1\n",
      "20\t59\t0.0\t1\t58 1\n",
      "22\t43\t0.0\t1\t41 2\n",
      "23\t28\t0.0\t1\t26 2\n",
      "24\t34\t0.0\t1\t33 1\n",
      "25\t24\t0.0\t1\t24\n",
      "26\t37\t0.0\t1\t37\n",
      "27\t47\t0.0\t1\t45 2\n",
      "28\t53\t0.0\t1\t51 2\n",
      "29\t58\t0.0\t1\t57 1\n",
      "30\t2\t0.0\t1\t0 2\n",
      "31\t32\t0.0\t1\t32\n",
      "32\t88\t0.0\t1\t87 1\n",
      "33\t1\t0.0\t1\t1\n",
      "34\t34\t0.0\t1\t31 3\n",
      "35\t37\t0.0\t1\t37\n",
      "36\t40\t0.0\t1\t39 1\n",
      "37\t49\t0.0\t1\t45 4\n",
      "38\t49\t0.0\t1\t49\n",
      "39\t74\t0.0\t1\t74\n",
      "40\t6\t0.0\t1\t5 1\n",
      "41\t37\t0.0\t1\t37\n",
      "42\t67\t0.0\t1\t63 4\n",
      "43\t42\t0.0\t1\t42\n",
      "44\t66\t0.0\t1\t61 5\n",
      "45\t224\t0.0\t1\t219 5\n",
      "46\t5\t0.0\t1\t4 1\n",
      "47\t18\t0.0\t1\t18\n",
      "48\t3\t0.0\t1\t3\n",
      "49\t22\t0.0\t1\t22\n",
      "50\t11\t0.0\t1\t11\n",
      "51\t44\t0.0\t1\t44\n",
      "52\t42\t0.0\t1\t41 1\n",
      "53\t47\t0.0\t1\t47\n",
      "54\t76\t0.0\t1\t75 1\n",
      "55\t37\t0.0\t1\t35 2\n",
      "56\t63\t0.0\t1\t62 1\n",
      "57\t181\t0.0\t1\t179 2\n",
      "58\t44\t0.0\t1\t42 2\n",
      "59\t21\t0.0\t1\t20 1\n",
      "60\t126\t0.0\t1\t123 3\n",
      "61\t26\t0.0\t1\t26\n",
      "62\t39\t0.0\t1\t37 2\n",
      "63\t198\t0.0\t1\t197 1\n",
      "64\t121\t0.0\t1\t118 3\n",
      "65\t12\t0.0\t1\t12\n",
      "66\t51\t0.0\t1\t51\n",
      "67\t77\t0.0\t1\t77\n",
      "68\t9\t0.0\t1\t9\n",
      "69\t46\t0.0\t1\t45 1\n",
      "70\t116\t0.0\t1\t116\n",
      "71\t50\t0.0\t1\t49 1\n",
      "72\t34\t0.0\t1\t33 1\n",
      "73\t45\t0.0\t1\t45\n",
      "74\t41\t0.0\t1\t38 3\n",
      "75\t45\t0.0\t1\t44 1\n",
      "76\t81\t0.0\t1\t79 2\n",
      "77\t47\t0.0\t1\t45 2\n",
      "78\t62\t0.0\t1\t61 1\n",
      "79\t57\t0.0\t1\t57\n",
      "80\t54\t0.0\t1\t53 1\n",
      "81\t80\t0.0\t1\t80\n",
      "82\t68\t0.0\t1\t66 2\n",
      "83\t76\t0.0\t1\t74 2\n",
      "84\t87\t0.0\t1\t87\n",
      "85\t86\t0.0\t1\t86\n",
      "86\t101\t0.0\t1\t101\n",
      "87\t85\t0.0\t1\t82 3\n",
      "88\t96\t0.0\t1\t94 2\n",
      "89\t93\t0.0\t1\t90 3\n",
      "90\t102\t0.0\t1\t102\n",
      "91\t86\t0.0\t1\t84 2\n",
      "92\t84\t0.0\t1\t84\n",
      "93\t105\t0.0\t1\t102 3\n",
      "94\t88\t0.0\t1\t88\n",
      "95\t104\t0.0\t1\t103 1\n",
      "96\t106\t0.0\t1\t105 1\n",
      "97\t82\t0.0\t1\t82\n",
      "98\t97\t0.0\t1\t94 3\n",
      "99\t104\t0.0\t1\t102 2\n",
      "100\t106\t0.0\t1\t105 1\n",
      "101\t91\t0.0\t1\t90 1\n",
      "102\t95\t0.0\t1\t94 1\n",
      "103\t127\t0.0\t1\t124 3\n",
      "104\t93\t0.0\t1\t93\n",
      "105\t108\t0.0\t1\t104 4\n",
      "106\t108\t0.0\t1\t107 1\n",
      "107\t92\t0.0\t1\t92\n",
      "108\t108\t0.0\t1\t106 2\n",
      "109\t84\t0.0\t1\t83 1\n",
      "110\t75\t0.0\t1\t75\n",
      "111\t60\t0.0\t1\t59 1\n",
      "112\t65\t0.0\t1\t65\n",
      "113\t60\t0.0\t1\t59 1\n",
      "114\t57\t0.0\t1\t55 2\n",
      "115\t47\t0.0\t1\t46 1\n",
      "116\t24\t0.0\t1\t24\n",
      "117\t13\t0.0\t1\t12 1\n",
      "118\t9\t0.0\t1\t9\n",
      "119\t4\t0.0\t1\t4\n",
      "120\t6\t0.0\t1\t6\n",
      "121\t5\t0.0\t1\t5\n",
      "122\t4\t0.0\t1\t4\n",
      "123\t2\t0.0\t1\t2\n",
      "125\t2\t0.0\t1\t2\n",
      "127\t1\t0.0\t1\t1\n",
      "129\t1\t0.0\t1\t1\n",
      "130\t1\t0.0\t1\t1\n",
      "132\t1\t0.0\t1\t1\n",
      "135\t1\t0.0\t1\t1\n",
      "136\t2\t0.0\t1\t2\n",
      "137\t2\t0.0\t1\t2\n",
      "138\t1\t0.0\t1\t1\n",
      "140\t2\t0.0\t1\t2\n",
      "143\t1\t0.0\t1\t1\n",
      "148\t3\t0.0\t1\t3\n",
      "150\t8\t0.0\t1\t7 1\n",
      "\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /scratch/hansetj1/script_testing/fastq_processing/raw_reads/RNA_sub_read2.fq\n",
      "=============================================\n",
      "10000 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files RNA_sub_read1_trimmed.fq and RNA_sub_read2_trimmed.fq\n",
      "Now renaming the output files\n",
      "\n",
      "ORIGINAL FILE 1: >>RNA_sub_read1_trimmed.fq<<\tRENAMING TO:>>test_set_RNA_R1_trimmed.fq<<\n",
      "ORIGINAL FILE 2: >>RNA_sub_read2_trimmed.fq<<\tRENAMING TO:>>test_set_RNA_R2_trimmed.fq<<\n",
      "file_1: test_set_RNA_R1_trimmed.fq, file_2: test_set_RNA_R2_trimmed.fq\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: test_set_RNA_R1_trimmed.fq and test_set_RNA_R2_trimmed.fq <<<<<\n",
      "Renaming the output files (AGAIN).\n",
      "Writing validated paired-end Read 1 reads to test_set_RNA_val_1.fq\n",
      "Writing validated paired-end Read 2 reads to test_set_RNA_val_2.fq\n",
      "\n",
      "Total number of sequences analysed: 10000\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 23 (0.23%)\n",
      "\n",
      "\n",
      "  >>> Now running FastQC on the validated data test_set_RNA_val_1.fq<<<\n",
      "\n",
      "Started analysis of test_set_RNA_val_1.fq\n",
      "Approx 10% complete for test_set_RNA_val_1.fq\n",
      "Approx 15% complete for test_set_RNA_val_1.fq\n",
      "Approx 30% complete for test_set_RNA_val_1.fq\n",
      "Approx 35% complete for test_set_RNA_val_1.fq\n",
      "Approx 45% complete for test_set_RNA_val_1.fq\n",
      "Approx 60% complete for test_set_RNA_val_1.fq\n",
      "Approx 70% complete for test_set_RNA_val_1.fq\n",
      "Approx 80% complete for test_set_RNA_val_1.fq\n",
      "Approx 90% complete for test_set_RNA_val_1.fq\n",
      "Analysis complete for test_set_RNA_val_1.fq\n",
      "\n",
      "  >>> Now running FastQC on the validated data test_set_RNA_val_2.fq<<<\n",
      "\n",
      "Started analysis of test_set_RNA_val_2.fq\n",
      "Approx 10% complete for test_set_RNA_val_2.fq\n",
      "Approx 20% complete for test_set_RNA_val_2.fq\n",
      "Approx 30% complete for test_set_RNA_val_2.fq\n",
      "Approx 35% complete for test_set_RNA_val_2.fq\n",
      "Approx 45% complete for test_set_RNA_val_2.fq\n",
      "Approx 60% complete for test_set_RNA_val_2.fq\n",
      "Approx 70% complete for test_set_RNA_val_2.fq\n",
      "Approx 80% complete for test_set_RNA_val_2.fq\n",
      "Approx 90% complete for test_set_RNA_val_2.fq\n",
      "Analysis complete for test_set_RNA_val_2.fq\n",
      "Deleting both intermediate output files test_set_RNA_R1_trimmed.fq and test_set_RNA_R2_trimmed.fq\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "\n",
      "***Mapping trimmed reads\n",
      "9977 reads; of these:\n",
      "  9977 (100.00%) were paired; of these:\n",
      "    6513 (65.28%) aligned concordantly 0 times\n",
      "    2868 (28.75%) aligned concordantly exactly 1 time\n",
      "    596 (5.97%) aligned concordantly >1 times\n",
      "34.72% overall alignment rate\n",
      "\n",
      "***Filtering mapped reads\n",
      "[bam_sort_core] merging from 0 files and 8 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 8 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 8 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 8 in-memory blocks...\n",
      "INFO\t2021-10-19 15:03:30\tMarkDuplicates\t\n",
      "\n",
      "********** NOTE: Picard's command line syntax is changing.\n",
      "**********\n",
      "********** For more information, please see:\n",
      "********** https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\n",
      "**********\n",
      "********** The command line looks like this in the new syntax:\n",
      "**********\n",
      "**********    MarkDuplicates -I /scratch/hansetj1/script_testing/fastq_processing/bams/test_set_RNA.filtered.pos-sorted.bam -O /scratch/hansetj1/script_testing/fastq_processing/bams/test_set_RNA.unique.unsorted.bam -M /scratch/hansetj1/script_testing/fastq_processing/qc/test_set_RNA_marked_dup_metrics-all.txt -REMOVE_DUPLICATES TRUE\n",
      "**********\n",
      "\n",
      "\n",
      "15:03:31.172 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/gpfs52/home/hansetj1/.conda/envs/jupyter/share/picard-2.26.3-0/picard.jar!/com/intel/gkl/native/libgkl_compression.so\n",
      "[Tue Oct 19 15:03:31 CDT 2021] MarkDuplicates INPUT=[/scratch/hansetj1/script_testing/fastq_processing/bams/test_set_RNA.filtered.pos-sorted.bam] OUTPUT=/scratch/hansetj1/script_testing/fastq_processing/bams/test_set_RNA.unique.unsorted.bam METRICS_FILE=/scratch/hansetj1/script_testing/fastq_processing/qc/test_set_RNA_marked_dup_metrics-all.txt REMOVE_DUPLICATES=true    MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP=50000 MAX_FILE_HANDLES_FOR_READ_ENDS_MAP=8000 SORTING_COLLECTION_SIZE_RATIO=0.25 TAG_DUPLICATE_SET_MEMBERS=false REMOVE_SEQUENCING_DUPLICATES=false TAGGING_POLICY=DontTag CLEAR_DT=true DUPLEX_UMI=false ADD_PG_TAG_TO_READS=true ASSUME_SORTED=false DUPLICATE_SCORING_STRATEGY=SUM_OF_BASE_QUALITIES PROGRAM_RECORD_ID=MarkDuplicates PROGRAM_GROUP_NAME=MarkDuplicates READ_NAME_REGEX=<optimized capture of last three ':' separated fields as numeric values> OPTICAL_DUPLICATE_PIXEL_DISTANCE=100 MAX_OPTICAL_DUPLICATE_SET_SIZE=300000 VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false GA4GH_CLIENT_SECRETS=client_secrets.json USE_JDK_DEFLATER=false USE_JDK_INFLATER=false\n",
      "[Tue Oct 19 15:03:31 CDT 2021] Executing as hansetj1@cn1288.vampire on Linux 3.10.0-1160.36.2.el7.x86_64 amd64; OpenJDK 64-Bit Server VM 1.8.0_302-b08; Deflater: Intel; Inflater: Intel; Provider GCS is not available; Picard version: 2.26.3\n",
      "INFO\t2021-10-19 15:03:31\tMarkDuplicates\tStart of doWork freeMemory: 1117100048; totalMemory: 1133510656; maxMemory: 16799760384\n",
      "INFO\t2021-10-19 15:03:31\tMarkDuplicates\tReading input file and constructing read end information.\n",
      "INFO\t2021-10-19 15:03:31\tMarkDuplicates\tWill retain up to 60868697 data points before spilling to disk.\n",
      "INFO\t2021-10-19 15:03:31\tMarkDuplicates\tRead 1684 records. 0 pairs never matched.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO\t2021-10-19 15:03:32\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 635769256; totalMemory: 1133510656; maxMemory: 16799760384\n",
      "INFO\t2021-10-19 15:03:32\tMarkDuplicates\tWill retain up to 524992512 duplicate indices before spilling to disk.\n",
      "INFO\t2021-10-19 15:03:34\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
      "INFO\t2021-10-19 15:03:34\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
      "INFO\t2021-10-19 15:03:34\tMarkDuplicates\tSorting list of duplicate records.\n",
      "INFO\t2021-10-19 15:03:35\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 1123629928; totalMemory: 5333581824; maxMemory: 16799760384\n",
      "INFO\t2021-10-19 15:03:35\tMarkDuplicates\tMarking 2 records as duplicates.\n",
      "INFO\t2021-10-19 15:03:35\tMarkDuplicates\tFound 0 optical duplicate clusters.\n",
      "INFO\t2021-10-19 15:03:36\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
      "INFO\t2021-10-19 15:03:36\tMarkDuplicates\tWriting complete. Closing input iterator.\n",
      "INFO\t2021-10-19 15:03:36\tMarkDuplicates\tDuplicate Index cleanup.\n",
      "INFO\t2021-10-19 15:03:36\tMarkDuplicates\tGetting Memory Stats.\n",
      "INFO\t2021-10-19 15:03:36\tMarkDuplicates\tBefore output close freeMemory: 5322929536; totalMemory: 5333581824; maxMemory: 16799760384\n",
      "INFO\t2021-10-19 15:03:36\tMarkDuplicates\tClosed outputs. Getting more Memory Stats.\n",
      "INFO\t2021-10-19 15:03:36\tMarkDuplicates\tAfter output close freeMemory: 5323010552; totalMemory: 5333581824; maxMemory: 16799760384\n",
      "[Tue Oct 19 15:03:36 CDT 2021] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 0.08 minutes.\n",
      "Runtime.totalMemory()=5333581824\n",
      "[bam_sort_core] merging from 0 files and 8 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 8 in-memory blocks...\n",
      "\n",
      "***Counting mapped reads\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "dir='/scratch/hansetj1/script_testing/fastq_processing'\n",
    "\n",
    "python3 ./fastq_processing/fastq_processing.py -a ${dir}/raw_reads/RNA_sub_read1.fq -b ${dir}/raw_reads/RNA_sub_read2.fq \\\n",
    "    -n test_set_RNA -e /home/hansetj1/hg38_encode_blacklist_ENCFF356LFX.bed -t ${dir}/trimmed_reads \\\n",
    "    -o ${dir}/bams -r ${dir}/qc -g /data/hodges_lab/hg38_genome/bt2/hg38"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "3e7379a4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1684 + 0 in total (QC-passed reads + QC-failed reads)\n",
      "1684 + 0 primary\n",
      "0 + 0 secondary\n",
      "0 + 0 supplementary\n",
      "0 + 0 duplicates\n",
      "0 + 0 primary duplicates\n",
      "1684 + 0 mapped (100.00% : N/A)\n",
      "1684 + 0 primary mapped (100.00% : N/A)\n",
      "1684 + 0 paired in sequencing\n",
      "842 + 0 read1\n",
      "842 + 0 read2\n",
      "1684 + 0 properly paired (100.00% : N/A)\n",
      "1684 + 0 with itself and mate mapped\n",
      "0 + 0 singletons (0.00% : N/A)\n",
      "0 + 0 with mate mapped to a different chr\n",
      "0 + 0 with mate mapped to a different chr (mapQ>=5)\n",
      "1682 + 0 in total (QC-passed reads + QC-failed reads)\n",
      "1682 + 0 primary\n",
      "0 + 0 secondary\n",
      "0 + 0 supplementary\n",
      "0 + 0 duplicates\n",
      "0 + 0 primary duplicates\n",
      "1682 + 0 mapped (100.00% : N/A)\n",
      "1682 + 0 primary mapped (100.00% : N/A)\n",
      "1682 + 0 paired in sequencing\n",
      "841 + 0 read1\n",
      "841 + 0 read2\n",
      "1682 + 0 properly paired (100.00% : N/A)\n",
      "1682 + 0 with itself and mate mapped\n",
      "0 + 0 singletons (0.00% : N/A)\n",
      "0 + 0 with mate mapped to a different chr\n",
      "0 + 0 with mate mapped to a different chr (mapQ>=5)\n"
     ]
    }
   ],
   "source": [
    "cat ${dir}/qc/test_set_RNA.filtered.pos-sorted.bam_flagstat-results.txt \\\n",
    "    ${dir}/qc/test_set_RNA.unique.pos-sorted.bam_flagstat-results.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb0e9b4e",
   "metadata": {},
   "source": [
    "## This script works!"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
