---
title: "Plotting Estimated Complexities and Fragment Distribution of ATAC-STARR Sequencing Data"
author: "Tyler Hansen"
date: "January 26, 2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Note: This analysis uses the results from both preseq and picard software packages. The code used to get these results is on accre in the Library_Complexity bin directory. 

## lc_extrap results
```{r message=FALSE, warning=FALSE}
library(tidyverse)
library(extrafont)
loadfonts(device = "win")
library(ggsci)
```

```{r Functions, message=FALSE}
read_lcExtrap <- function(input_file, condition, library_type){
  #This function reads in the lcextrap txt file and assigns descriptors of the data to new columns for use in plotting. 
  #condition example is "GM12878inGM12878", library_type example is "DNA_Rep1"
  return(read_tsv(input_file) %>% mutate(condition=condition, Type=library_type))
}

read_ISM <- function(input_file, condition, library_type){
  #This function reads in the lcextrap txt file and assigns descriptors of the data to new columns for use in plotting. 
  #condition example is "GM12878inGM12878", library_type example is "DNA_Rep1"
  tsv <- read_tsv(input_file, skip = 11) %>% mutate(condition=condition, Type=library_type)
  return(mutate(tsv, cpm=(All_Reads.fr_count/sum(tsv$All_Reads.fr_count))*1e6))
}
```

```{r, Plot lc_extrap results, message=FALSE}
#Read in results and assign to unique data.frame objects. 
setwd("C:/Users/tyler/Dropbox (VU Basic Sciences)/Hodges Lab/Results and Data/Bioinformatics/ATAC-STARR_analysis/3_Accre_Local/ATAC-STARR/results/Library_Complexity")
GGD1 <- read_lcExtrap(input_file = 'lc-extrap/GM12878inGM12878_DNA_Rep1_lc-extrap-results.txt', condition = "GM12878inGM12878", library_type = "DNA Rep1")
GGD2 <- read_lcExtrap(input_file = 'lc-extrap/GM12878inGM12878_DNA_Rep2_lc-extrap-results.txt', condition = "GM12878inGM12878", library_type = "DNA Rep2")
GGR1 <- read_lcExtrap(input_file = 'lc-extrap/GM12878inGM12878_RNA_Rep1_lc-extrap-results.txt', condition = "GM12878inGM12878", library_type = "RNA Rep1")
GGR2 <- read_lcExtrap(input_file = 'lc-extrap/GM12878inGM12878_RNA_Rep2_lc-extrap-results.txt', condition = "GM12878inGM12878", library_type = "RNA Rep2")

GLD1 <- read_lcExtrap(input_file = 'lc-extrap/GM12878inLCL8664_DNA_Rep1_lc-extrap-results.txt', condition = "GM12878inLCL8664", library_type = "DNA Rep1")
GLD2 <- read_lcExtrap(input_file = 'lc-extrap/GM12878inLCL8664_DNA_Rep2_lc-extrap-results.txt', condition = "GM12878inLCL8664", library_type = "DNA Rep2")
GLR1 <- read_lcExtrap(input_file = 'lc-extrap/GM12878inLCL8664_RNA_Rep1_lc-extrap-results.txt', condition = "GM12878inLCL8664", library_type = "RNA Rep1")
GLR2 <- read_lcExtrap(input_file = 'lc-extrap/GM12878inLCL8664_RNA_Rep2_lc-extrap-results.txt', condition = "GM12878inLCL8664", library_type = "RNA Rep2")

LGD1 <- read_lcExtrap(input_file = 'lc-extrap/LCL8664inGM12878_DNA_Rep1_lc-extrap-results.txt', condition = "LCL8664inGM12878", library_type = "DNA Rep1")
LGD2 <- read_lcExtrap(input_file = 'lc-extrap/LCL8664inGM12878_DNA_Rep2_lc-extrap-results.txt', condition = "LCL8664inGM12878", library_type = "DNA Rep2")
LGR1 <- read_lcExtrap(input_file = 'lc-extrap/LCL8664inGM12878_RNA_Rep1_lc-extrap-results.txt', condition = "LCL8664inGM12878", library_type = "RNA Rep1")
LGR2 <- read_lcExtrap(input_file = 'lc-extrap/LCL8664inGM12878_RNA_Rep2_lc-extrap-results.txt', condition = "LCL8664inGM12878", library_type = "RNA Rep2")

LLD1 <- read_lcExtrap(input_file = 'lc-extrap/LCL8664inLCL8664_DNA_Rep1_lc-extrap-results.txt', condition = "LCL8664inLCL8664", library_type = "DNA Rep1")
LLD2 <- read_lcExtrap(input_file = 'lc-extrap/LCL8664inLCL8664_DNA_Rep2_lc-extrap-results.txt', condition = "LCL8664inLCL8664", library_type = "DNA Rep2")
LLR1 <- read_lcExtrap(input_file = 'lc-extrap/LCL8664inLCL8664_RNA_Rep1_lc-extrap-results.txt', condition = "LCL8664inLCL8664", library_type = "RNA Rep1")
LLR2 <- read_lcExtrap(input_file = 'lc-extrap/LCL8664inLCL8664_RNA_Rep2_lc-extrap-results.txt', condition = "LCL8664inLCL8664", library_type = "RNA Rep2")

#Merge into one dataframe. 
lcExtrap_all <- bind_rows(GGD1, GGD2, GGR1, GGR2, GLD1, GLD2, GLR1, GLR2, LGD1, LGD2, LGR1, LGR2, LLD1, LLD2, LLR1, LLR2)

#Plot:
lcExtrap_plot <- ggplot(lcExtrap_all) + 
  geom_ribbon(aes(x=TOTAL_READS/10^6, ymin=LOWER_0.95CI/10^6, ymax=UPPER_0.95CI/10^6, fill=Type)) +
  geom_line(aes(x=TOTAL_READS/10^6, y=EXPECTED_DISTINCT/10^6, colour=Type), size=1, linetype="dashed") +
  xlab("Sequenced reads (M)") + 
  ylab("Distinct reads (M)") +
  ggtitle("Estimated Complexity of ATAC-STARR-seq Libraries") +
  theme_bw(base_size = 8, base_family = "Arial") +
  theme(plot.title = element_text(hjust = 0.5, size = 12, face = "bold")) +
  coord_cartesian(xlim = c(0, 1000), ylim = c(0,100)) +
  scale_color_npg() +
  scale_fill_npg(alpha = 0.25) +
  facet_grid(cols = vars(condition)) +
  theme(strip.text.x = element_text(size = 8, family = "Arial", face = "italic", color = "black"), strip.background = element_rect(fill = "lightsteelblue"))

lcExtrap_plot

setwd("C:/Users/tyler/Dropbox (VU Basic Sciences)/Hodges Lab/Results and Data/Bioinformatics/ATAC-STARR_analysis/3_Accre_Local/ATAC-STARR/results/Library_Complexity")
ggsave(filename = "lc-extrap_all-samples.pdf", plot = lcExtrap_plot, units = "in", width = 8, height = 4, device = cairo_pdf)
```

```{r, Plot InsertSizeMetrics results, message=FALSE}
#Read in results and assign to unique data.frame objects. 
setwd("C:/Users/tyler/Dropbox (VU Basic Sciences)/Hodges Lab/Results and Data/Bioinformatics/ATAC-STARR_analysis/3_Accre_Local/ATAC-STARR/results/Library_Complexity")
GGD1 <- read_ISM(input_file = 'Picard-CISM/GM12878inGM12878_DNA_Rep1_picard-CISM.txt', condition = "GM12878inGM12878", library_type = "DNA Rep1")
GGD2 <- read_ISM(input_file = 'Picard-CISM/GM12878inGM12878_DNA_Rep2_picard-CISM.txt', condition = "GM12878inGM12878", library_type = "DNA Rep2")
GGR1 <- read_ISM(input_file = 'Picard-CISM/GM12878inGM12878_RNA_Rep1_picard-CISM.txt', condition = "GM12878inGM12878", library_type = "RNA Rep1")
GGR2 <- read_ISM(input_file = 'Picard-CISM/GM12878inGM12878_RNA_Rep2_picard-CISM.txt', condition = "GM12878inGM12878", library_type = "RNA Rep2")

GLD1 <- read_ISM(input_file = 'Picard-CISM/GM12878inLCL8664_DNA_Rep1_picard-CISM.txt', condition = "GM12878inLCL8664", library_type = "DNA Rep1")
GLD2 <- read_ISM(input_file = 'Picard-CISM/GM12878inLCL8664_DNA_Rep2_picard-CISM.txt', condition = "GM12878inLCL8664", library_type = "DNA Rep2")
GLR1 <- read_ISM(input_file = 'Picard-CISM/GM12878inLCL8664_RNA_Rep1_picard-CISM.txt', condition = "GM12878inLCL8664", library_type = "RNA Rep1")
GLR2 <- read_ISM(input_file = 'Picard-CISM/GM12878inLCL8664_RNA_Rep2_picard-CISM.txt', condition = "GM12878inLCL8664", library_type = "RNA Rep2")

LGD1 <- read_ISM(input_file = 'Picard-CISM/LCL8664inGM12878_DNA_Rep1_picard-CISM.txt', condition = "LCL8664inGM12878", library_type = "DNA Rep1")
LGD2 <- read_ISM(input_file = 'Picard-CISM/LCL8664inGM12878_DNA_Rep2_picard-CISM.txt', condition = "LCL8664inGM12878", library_type = "DNA Rep2")
LGR1 <- read_ISM(input_file = 'Picard-CISM/LCL8664inGM12878_RNA_Rep1_picard-CISM.txt', condition = "LCL8664inGM12878", library_type = "RNA Rep1")
LGR2 <- read_ISM(input_file = 'Picard-CISM/LCL8664inGM12878_RNA_Rep2_picard-CISM.txt', condition = "LCL8664inGM12878", library_type = "RNA Rep2")

LLD1 <- read_ISM(input_file = 'Picard-CISM/LCL8664inLCL8664_DNA_Rep1_picard-CISM.txt', condition = "LCL8664inLCL8664", library_type = "DNA Rep1")
LLD2 <- read_ISM(input_file = 'Picard-CISM/LCL8664inLCL8664_DNA_Rep2_picard-CISM.txt', condition = "LCL8664inLCL8664", library_type = "DNA Rep2")
LLR1 <- read_ISM(input_file = 'Picard-CISM/LCL8664inLCL8664_RNA_Rep1_picard-CISM.txt', condition = "LCL8664inLCL8664", library_type = "RNA Rep1")
LLR2 <- read_ISM(input_file = 'Picard-CISM/LCL8664inLCL8664_RNA_Rep2_picard-CISM.txt', condition = "LCL8664inLCL8664", library_type = "RNA Rep2")

#Merge into one dataframe. 
ISM_all <- bind_rows(GGD1, GGD2, GGR1, GGR2, GLD1, GLD2, GLR1, GLR2, LGD1, LGD2, LGR1, LGR2, LLD1, LLD2, LLR1, LLR2)

#Plot:
ISM_plot <- ggplot(ISM_all, aes(x=insert_size, y=cpm)) +
  geom_line(aes(colour=Type), size = 0.5) +
  facet_grid(rows = vars(Type), cols = vars(condition)) +
  coord_cartesian(x=c(0,500)) +
  scale_x_continuous(breaks = seq(0, 500, by = 100)) +
  scale_color_manual(values = c("black", "black", "black", "black")) +
  xlab("Insert Size (bp)") + 
  ylab("Counts per Million") +
  ggtitle("Insert Size Distribution of ATAC-STARR-seq Libraries") +
  theme_bw(base_size = 8, base_family = "Arial") +
  theme(plot.title = element_text(hjust = 0.5, size = 12, face = "bold"), 
        strip.text.x = element_text(size = 8, family = "Arial", face = "italic", color = "black"),
        strip.text.y = element_text(size = 8, family = "Arial", face = "plain", color = "black"),
        strip.background = element_rect(fill = "lightgray"), 
        legend.position = "none")

ISM_plot

setwd("C:/Users/tyler/Dropbox (VU Basic Sciences)/Hodges Lab/Results and Data/Bioinformatics/ATAC-STARR_analysis/3_Accre_Local/ATAC-STARR/results/Library_Complexity")
ggsave(filename = "InsertSizeMetrics_all-samples.pdf", plot = ISM_plot, units = "in", width = 7, height = 4.8, device = cairo_pdf)
```
