---
title: "clustering_part2"
author: "Tyler Hansen"
date: "6/17/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r Libraries}
library(tidyverse)
library(pheatmap)
library(viridis)
set.seed(123)
library(extrafont)
loadfonts(device = "win")
library(ggsci)
library(ChIPseeker)
library(clusterProfiler)
library("GenomicRanges")
library(org.Hs.eg.db)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(ReactomePA)
```

```{r Clustering by TF Binding - Active}
#Read in results and sort by peak_ID for all
IRF <- read_tsv("C:/Users/tyler/Dropbox (VU Basic Sciences)/Hodges Lab/Results and Data/Bioinformatics/ATAC-STARR_analysis/3_Accre_Local/ATAC-STARR/results/multiomic/Cluster/IRF4_MA1419.1-in-GM12878inGM12878_active.bed", skip = 1, col_names = c("chr","start","end","Peak_ID","Score","Strand","IRF")) %>% arrange(Peak_ID)

SPI <- read_tsv("C:/Users/tyler/Dropbox (VU Basic Sciences)/Hodges Lab/Results and Data/Bioinformatics/ATAC-STARR_analysis/3_Accre_Local/ATAC-STARR/results/multiomic/Cluster/SPI1_MA0080.5-in-GM12878inGM12878_active.bed", skip = 1, col_names = c("chr","start","end","Peak_ID","Score","Strand","SPI")) %>% arrange(Peak_ID)

CREB <- read_tsv("C:/Users/tyler/Dropbox (VU Basic Sciences)/Hodges Lab/Results and Data/Bioinformatics/ATAC-STARR_analysis/3_Accre_Local/ATAC-STARR/results/multiomic/Cluster/JUNB_MA0490.2-in-GM12878inGM12878_active.bed", skip = 1, col_names = c("chr","start","end","Peak_ID","Score","Strand","CREB")) %>% arrange(Peak_ID)

NFKB <- read_tsv("C:/Users/tyler/Dropbox (VU Basic Sciences)/Hodges Lab/Results and Data/Bioinformatics/ATAC-STARR_analysis/3_Accre_Local/ATAC-STARR/results/multiomic/Cluster/NFKB1_MA0105.4-in-GM12878inGM12878_active.bed", skip = 1, col_names = c("chr","start","end","Peak_ID","Score","Strand","NFKB")) %>% arrange(Peak_ID)

AP1 <- read_tsv("C:/Users/tyler/Dropbox (VU Basic Sciences)/Hodges Lab/Results and Data/Bioinformatics/ATAC-STARR_analysis/3_Accre_Local/ATAC-STARR/results/multiomic/Cluster/BACH2_MA1101.2-in-GM12878inGM12878_active.bed", skip = 1, col_names = c("chr","start","end","Peak_ID","Score","Strand","AP1")) %>% arrange(Peak_ID) 

ETS <- read_tsv("C:/Users/tyler/Dropbox (VU Basic Sciences)/Hodges Lab/Results and Data/Bioinformatics/ATAC-STARR_analysis/3_Accre_Local/ATAC-STARR/results/multiomic/Cluster/ELK1_MA0028.2-in-GM12878inGM12878_active.bed", skip = 1, col_names = c("chr","start","end","Peak_ID","Score","Strand","ETS")) %>% arrange(Peak_ID)

#cbind each TF together. Since Peak_ID order is the same for all, only need to join TF column. 
active <- cbind(IRF, SPI = SPI$SPI, CREB = CREB$CREB, NFKB = NFKB$NFKB, AP1 = AP1$AP1, ETS = ETS$ETS)

#Convert to matrix
active_matrix <- column_to_rownames(active, var = "Peak_ID") %>% dplyr::select(-chr, -start, -end, -Score, -Strand)

#Remove all regions without a footprint. 
active_matrix_subset <- filter(active_matrix, rowSums(active_matrix) != 0)

#Heatmap
heatmap <- pheatmap(active_matrix_subset, show_rownames = FALSE, cluster_cols=TRUE, cluster_rows=TRUE, cellwidth = 50, cutree_rows = 6, display_numbers = FALSE, legend = FALSE, clustering_distance_rows = 'binary', clustering_distance_cols = 'binary')

#Assign cluster number to each region. cbind cluster to the input matrix and then left_join with df to get genomic coordinates.  
active_cluster <- cbind(active_matrix_subset, Cluster = cutree(heatmap$tree_row,k = 6)) %>% rownames_to_column(var = "Peak_ID") %>% left_join(active, by = "Peak_ID") %>% dplyr::select(chr, start, end, Peak_ID, Cluster)

#split into 6 clusters. 
ETS_clust <- filter(active_cluster, Cluster == 1)
IRF_clust <- filter(active_cluster, Cluster == 2)
SPI_clust <- filter(active_cluster, Cluster == 3)
AP1_CREB_clust <- filter(active_cluster, Cluster == 4)
NFKB_clust <- filter(active_cluster, Cluster == 5)
CREB_clust <- filter(active_cluster, Cluster == 6)

#make gRanges object and annotate peaks to assign nearest neighbor. 
ETS_anno <- annotatePeak(with(ETS_clust, GRanges(chr, IRanges(start, end))), tssRegion = c(-2000, 1000), TxDb = TxDb.Hsapiens.UCSC.hg38.knownGene, annoDb = "org.Hs.eg.db")
IRF_anno <- annotatePeak(with(IRF_clust, GRanges(chr, IRanges(start, end))), tssRegion = c(-2000, 1000), TxDb = TxDb.Hsapiens.UCSC.hg38.knownGene, annoDb = "org.Hs.eg.db")
SPI_anno <- annotatePeak(with(SPI_clust, GRanges(chr, IRanges(start, end))), tssRegion = c(-2000, 1000), TxDb = TxDb.Hsapiens.UCSC.hg38.knownGene, annoDb = "org.Hs.eg.db")
AP1_CREB_anno <- annotatePeak(with(AP1_CREB_clust, GRanges(chr, IRanges(start, end))), tssRegion = c(-2000, 1000), TxDb = TxDb.Hsapiens.UCSC.hg38.knownGene, annoDb = "org.Hs.eg.db")
NFKB_anno <- annotatePeak(with(NFKB_clust, GRanges(chr, IRanges(start, end))), tssRegion = c(-2000, 1000), TxDb = TxDb.Hsapiens.UCSC.hg38.knownGene, annoDb = "org.Hs.eg.db")
CREB_anno <- annotatePeak(with(CREB_clust, GRanges(chr, IRanges(start, end))), tssRegion = c(-2000, 1000), TxDb = TxDb.Hsapiens.UCSC.hg38.knownGene, annoDb = "org.Hs.eg.db")

#perform reactome pathway enrichment analysis of nearest neighbor gene sets. Make list of genes first. Then run on all simultaneously with clusterProfiler
gene_list <- list(ETS = as.data.frame(ETS_anno)$geneId, IRF = as.data.frame(IRF_anno)$geneId, SPI = as.data.frame(SPI_anno)$geneId, `AP1&CREB` = as.data.frame(AP1_CREB_anno)$geneId, NFKB = as.data.frame(NFKB_anno)$geneId, CREB = as.data.frame(CREB_anno)$geneId)
reactome_results <- compareCluster(gene_list, fun = "enrichPathway", pvalueCutoff = 0.2)
dotplot(reactome_results, showCategory = 3) + theme_bw() + theme(legend.position = "bottom")

#Rather than showing dotplot, which is clunky, make a column plot of p.adj for select pathways. 
res_df <- as.data.frame(reactome_results) 
filtered_df <- dplyr::filter(res_df, Description == c("Negative regulation of MAPK pathway"))
filtered_df <- dplyr::filter(res_df, Description == c("Toll Like Receptor 3 (TLR3) Cascade")) %>% bind_rows(filtered_df)
filtered_df <- dplyr::filter(res_df, Description == "Translation") %>% bind_rows(filtered_df)
filtered_df <- dplyr::filter(res_df, Description == "Interferon alpha/beta signaling") %>% bind_rows(filtered_df)
filtered_df <- dplyr::filter(res_df, Description == "Signaling by Interleukins") %>% bind_rows(filtered_df)
filtered_df <- dplyr::filter(res_df, Description == "rRNA processing") %>% bind_rows(filtered_df)
filtered_df <- dplyr::filter(res_df, Description == "Costimulation by the CD28 family") %>% bind_rows(filtered_df)
filtered_df <- dplyr::filter(res_df, Description == "Antiviral mechanism by IFN-stimulated genes") %>% bind_rows(filtered_df)
filtered_df <- dplyr::filter(res_df, Description == "Regulation by c-FLIP") %>% bind_rows(filtered_df)
filtered_df <- dplyr::filter(res_df, Description == "CASP8 activity is inhibited") %>% bind_rows(filtered_df)

filtered_df$Cluster <- factor(filtered_df$Cluster, levels = c("CREB", "NFKB", "AP1&CREB", "SPI", "IRF", "ETS")) 

p <- ggplot(filtered_df, aes(y=Description, x=-log10(p.adjust))) + 
  geom_col(orientation = "y", aes(fill = Cluster)) + 
  facet_grid(Cluster~., space = "free_y", scales = "free_y") + 
  theme_classic(base_size = 10, base_family = "Arial") + 
  scale_fill_brewer(palette = "Purples") +
  theme(strip.text = element_blank(), axis.title.y = element_blank(), legend.position = "none")

ggsave(p, filename = "results/multiomic/Cluster/reactome-bar.pdf", device = cairo_pdf, width = 6, height = 6)
```

```{r Clustering by TF Binding - Silent}
#Read in results and sort by peak_ID for all
CTCF <- read_tsv("C:/Users/tyler/Dropbox (VU Basic Sciences)/Hodges Lab/Results and Data/Bioinformatics/ATAC-STARR_analysis/3_Accre_Local/ATAC-STARR/results/multiomic/Cluster/CTCF_MA0139.1-in-GM12878inGM12878_silent.bed", skip = 1, col_names = c("chr","start","end","Peak_ID","Score","Strand","CTCF")) %>% arrange(Peak_ID)

SP1 <- read_tsv("C:/Users/tyler/Dropbox (VU Basic Sciences)/Hodges Lab/Results and Data/Bioinformatics/ATAC-STARR_analysis/3_Accre_Local/ATAC-STARR/results/multiomic/Cluster/SP1_MA0079.4-in-GM12878inGM12878_silent.bed", skip = 1, col_names = c("chr","start","end","Peak_ID","Score","Strand","SP1")) %>% arrange(Peak_ID)

ZBTB7A <- read_tsv("C:/Users/tyler/Dropbox (VU Basic Sciences)/Hodges Lab/Results and Data/Bioinformatics/ATAC-STARR_analysis/3_Accre_Local/ATAC-STARR/results/multiomic/Cluster/ZBTB7A_MA0750.2-in-GM12878inGM12878_silent.bed", skip = 1, col_names = c("chr","start","end","Peak_ID","Score","Strand","ZBTB7A")) %>% arrange(Peak_ID)

NFYA <- read_tsv("C:/Users/tyler/Dropbox (VU Basic Sciences)/Hodges Lab/Results and Data/Bioinformatics/ATAC-STARR_analysis/3_Accre_Local/ATAC-STARR/results/multiomic/Cluster/NFYA_MA0060.3-in-GM12878inGM12878_silent.bed", skip = 1, col_names = c("chr","start","end","Peak_ID","Score","Strand","NFYA")) %>% arrange(Peak_ID)

#cbind each TF together. Since Peak_ID order is the same for all, only need to join TF column. 
silent <- cbind(CTCF, SP1 = SP1$SP1, ZBTB7A = ZBTB7A$ZBTB7A, NFYA = NFYA$NFYA)

#Convert to matrix
silent_matrix <- column_to_rownames(silent, var = "Peak_ID") %>% dplyr::select(-chr, -start, -end, -Score, -Strand)

#Remove all regions without a footprint. 
silent_matrix_subset <- filter(silent_matrix, rowSums(silent_matrix) != 0)

#Heatmap
heatmap <- pheatmap(silent_matrix_subset, show_rownames = FALSE, cluster_cols=TRUE, cluster_rows=TRUE, cellwidth = 50, cutree_rows = 4, display_numbers = FALSE, legend = FALSE, clustering_distance_rows = 'binary', clustering_distance_cols = 'binary')

#Assign cluster number to each region. cbind cluster to the input matrix and then left_join with df to get genomic coordinates.  
silent_cluster <- cbind(silent_matrix_subset, Cluster = cutree(heatmap$tree_row,k = 4)) %>% rownames_to_column(var = "Peak_ID") %>% left_join(silent, by = "Peak_ID") %>% dplyr::select(chr, start, end, Peak_ID, Cluster)

#split into 6 clusters. 
ZBTB7A_clust <- filter(silent_cluster, Cluster == 1)
CTCF_clust <- filter(silent_cluster, Cluster == 2)
SP1_clust <- filter(silent_cluster, Cluster == 3)
NFYA_clust <- filter(silent_cluster, Cluster == 4)

#make gRanges object and annotate peaks to assign nearest neighbor. 
ZBTB7A_anno <- annotatePeak(with(ZBTB7A_clust, GRanges(chr, IRanges(start, end))), tssRegion = c(-2000, 1000), TxDb = TxDb.Hsapiens.UCSC.hg38.knownGene, annoDb = "org.Hs.eg.db")
CTCF_anno <- annotatePeak(with(CTCF_clust, GRanges(chr, IRanges(start, end))), tssRegion = c(-2000, 1000), TxDb = TxDb.Hsapiens.UCSC.hg38.knownGene, annoDb = "org.Hs.eg.db")
SP1_anno <- annotatePeak(with(SP1_clust, GRanges(chr, IRanges(start, end))), tssRegion = c(-2000, 1000), TxDb = TxDb.Hsapiens.UCSC.hg38.knownGene, annoDb = "org.Hs.eg.db")
NFYA_anno <- annotatePeak(with(NFYA_clust, GRanges(chr, IRanges(start, end))), tssRegion = c(-2000, 1000), TxDb = TxDb.Hsapiens.UCSC.hg38.knownGene, annoDb = "org.Hs.eg.db")

#perform reactome pathway enrichment analysis of nearest neighbor gene sets. Make list of genes first. Then run on all simultaneously with clusterProfiler
silent_gene_list <- list(ZBTB7A = as.data.frame(ZBTB7A_anno)$geneId, CTCF = as.data.frame(CTCF_anno)$geneId, SP1 = as.data.frame(SP1_anno)$geneId, NFYA = as.data.frame(NFYA_anno)$geneId)
silent_reactome_results <- compareCluster(silent_gene_list, fun = "enrichPathway", pvalueCutoff = 0.2)
dotplot(silent_reactome_results) + theme_bw() + theme(legend.position = "bottom")

#Rather than showing dotplot, which is clunky, make a column plot of p.adj for select pathways. 
res_df <- as.data.frame(silent_reactome_results) 
filtered_df <- dplyr::filter(res_df, Description == c("HIV Life Cycle"))
filtered_df <- dplyr::filter(res_df, Description == c("Interleukin-3, Interleukin-5 and GM-CSF signaling")) %>% bind_rows(filtered_df)
filtered_df <- dplyr::filter(res_df, Description == "Translation") %>% bind_rows(filtered_df)
filtered_df <- dplyr::filter(res_df, Description == "Regulation of signaling by CBL") %>% bind_rows(filtered_df)
filtered_df <- dplyr::filter(res_df, Description == "M Phase") %>% bind_rows(filtered_df)
filtered_df <- dplyr::filter(res_df, Description == "rRNA processing") %>% bind_rows(filtered_df)
filtered_df <- dplyr::filter(res_df, Description == "Cellular response to stress") %>% bind_rows(filtered_df)
filtered_df <- dplyr::filter(res_df, Description == "The citric (TCA) cycle and respitory electron transport") %>% bind_rows(filtered_df)
filtered_df <- dplyr::filter(res_df, Description == "Meiotic synapsis") %>% bind_rows(filtered_df)
filtered_df <- dplyr::filter(res_df, Description == "RUNX1 regulates genes involved in megakaryocyte differentiation and platelet function") %>% bind_rows(filtered_df)

filtered_df$Cluster <- factor(filtered_df$Cluster, levels = c("NFYA", "SP1", "CTCF", "ZBTB7A")) 

p <- ggplot(filtered_df, aes(y=Description, x=-log10(p.adjust))) + 
  geom_col(orientation = "y", aes(fill = Cluster)) + 
  facet_grid(Cluster~., space = "free_y", scales = "free_y") + 
  theme_classic(base_size = 10, base_family = "Arial") + 
  scale_fill_brewer(palette = "Purples") +
  theme(strip.text = element_blank(), axis.title.y = element_blank(), legend.position = "none")

ggsave(p, filename = "results/multiomic/Cluster/reactome-bar_silent.pdf", device = cairo_pdf, width = 8, height = 5)
```
