---
title: "Signal_Visualization_Sushi"
author: "Tyler Hansen"
date: "5/4/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, Libraries}
library(Sushi)
library(tidyverse)
```

```{r}
AS_ATAC <- read.delim("C:/Users/tyler/Dropbox (VU Basic Sciences)/Hodges Lab/Results and Data/Bioinformatics/ATAC-STARR_analysis/3_Accre_Local/ATAC-STARR/results/accessibility/signal_files/GM12878inGM12878_DNA_no-dup.bedGraph", header = FALSE)

buenrostro <- read.delim("C:/Users/tyler/Dropbox (VU Basic Sciences)/Hodges Lab/Results and Data/Bioinformatics/ATAC-STARR_analysis/3_Accre_Local/ATAC-STARR/results/accessibility/signal_files/GM12878_ATAC-seq_buenrostro_no-dup.bedGraph", header = FALSE)
```

```{r}
dev.off()
chrom = "chr14"
chromstart = 105037802
chromend = 105080491
layout(matrix(c(1,1,
                2,2), ncol = 2, byrow = TRUE))
par(mar=c(3,4,1,1))

plotBedgraph(AS_ATAC, chrom,chromstart,chromend,transparency=1,color="steelblue")
axis(side=2,las=2,tcl=.2)
labelgenome(chrom,chromstart,chromend,side=1,n=3,scale="Kb")

plotBedgraph(buenrostro, chrom, chromstart, chromend,transparency=1,color="forestgreen")
axis(side=2,las=2,tcl=.2)
labelgenome(chrom,chromstart,chromend,side=1,n=3,scale="Kb")
```

